{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Peptide Chef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Created by: Tyler T. Cooper, PhD (2021)\n",
    "# Tools to Analyze Peptides and Amino Acid Compositions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Data Handling\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import ExcelWriter\n",
    "import openpyxl\n",
    "import re\n",
    "import pyteomics\n",
    "from pyteomics import fasta, parser, electrochem, mass\n",
    "from itertools import combinations\n",
    "from urllib.request import urlretrieve\n",
    "import gzip\n",
    "import itertools as it\n",
    "from collections import deque\n",
    "\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "\n",
    "#Figure Generation\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib import offsetbox\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "from matplotlib.ticker import NullFormatter\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from matplotlib.tri import Triangulation\n",
    "from mpl_toolkits.mplot3d import axes3d\n",
    "from IPython.display import Image, display\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "import glob\n",
    "import bioinfokit\n",
    "from bioinfokit import analys, visuz\n",
    "from wordcloud import WordCloud\n",
    "\n",
    "#Venn Diagrams\n",
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "from matplotlib_venn import venn3, venn3_circles\n",
    "%matplotlib inline\n",
    "\n",
    "cmap = 'PRGn'\n",
    "fmt='eps'\n",
    "dpi=600"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Fasta File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "url=\"https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/UP000005640_9606.fasta.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CookBook(Species=None,homebrew=False, takeout=True,url=None,measure=False,target=None):\n",
    "    #Use to import fasta.gz files directly from uniprot (Takeout) with a provided url (Default is Human) or from local folder using homebrew hyperparameter.\n",
    "    # Takeout is default, Takeout must be switched to False for Homebrew to be True.\n",
    "    # accepts Uniprot format\n",
    "    ingredients=list()\n",
    "    if url == None:\n",
    "        url=\"https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/UP000005640_9606.fasta.gz\"\n",
    "    if (homebrew == True) & (takeout==True):\n",
    "        print(\"Only one method (homebrew/takeout) can == True...\")\n",
    "        print (\"Takeout will be automatically selected...\")\n",
    "    if takeout is True:\n",
    "        homebrew=False\n",
    "        print('Downloading the FASTA file from url...')\n",
    "        urlretrieve(url,'temp.fasta.gz')\n",
    "        print('Unzipping...')\n",
    "        with gzip.open('temp.fasta.gz', mode='rt') as gzfile:\n",
    "            for info, contents in fasta.FASTA(gzfile):\n",
    "                taste=list((info,contents))\n",
    "                ingredients.append(taste)\n",
    "                recipie=pd.DataFrame(ingredients,columns=['ID','Peptide'])\n",
    "        print(\"Takeout is Done!\")\n",
    "    if homebrew is True:\n",
    "        if Species == None:\n",
    "            print(\"Provide Species identification. This will be the name of the local FASTA file.\")\n",
    "        print(\"Downloading the FASTA file from local flle...\")\n",
    "        book = str(Species+\".fasta\")\n",
    "        recipie=pd.DataFrame()\n",
    "        print(\"Serving up a homebrew...\")\n",
    "        with fasta.read(book) as menu:\n",
    "            for info, contents in menu:\n",
    "                taste=list((info,contents))\n",
    "                ingredients.append(taste)\n",
    "                recipie=pd.DataFrame(ingredients,columns=['ID','Peptide'])\n",
    "        print(\"Homebrow is Done!\")\n",
    "    recipie[['db', 'UniprotID','ID2']] = recipie['ID'].str.split('|', 2, expand=True)\n",
    "    recipie[['Gene','Identification']] = recipie['ID2'].str.split('_', 1, expand=True)\n",
    "    recipie.drop(columns=['ID', 'ID2',\"db\"], inplace=True)\n",
    "    if measure == True:\n",
    "        recipie[\"Protein_Length\"]=recipie[target].str.len()\n",
    "    print(\"Here ya go boss!\")\n",
    "    return(recipie)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading the FASTA file from url...\n",
      "Unzipping...\n",
      "Takeout is Done!\n",
      "Here ya go boss!\n"
     ]
    }
   ],
   "source": [
    "menu=CookBook(homebrew=False,takeout=True,url=url, measure=True, target=\"Peptide\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20610"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(menu.Gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Identification</th>\n",
       "      <th>Protein_Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7421</th>\n",
       "      <td>MTTSQKHRDFVAEPMGEKPVGSLAGIGEVLGKKLEERGFDKAYVVL...</td>\n",
       "      <td>O75531</td>\n",
       "      <td>BAF</td>\n",
       "      <td>HUMAN Barrier-to-autointegration factor OS=Hom...</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14325</th>\n",
       "      <td>MLRAAARFGPRLGRRLLSAAATQAVPAPNQQPEVFCNQIFINNEWH...</td>\n",
       "      <td>P05091</td>\n",
       "      <td>ALDH2</td>\n",
       "      <td>HUMAN Aldehyde dehydrogenase, mitochondrial OS...</td>\n",
       "      <td>517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20606</th>\n",
       "      <td>MAALGTVLFTGVRRLHCSVAAWAGGQWRLQQGLAANPSGYGPLTEL...</td>\n",
       "      <td>Q86TS9</td>\n",
       "      <td>RM52</td>\n",
       "      <td>HUMAN 39S ribosomal protein L52, mitochondrial...</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Peptide UniprotID   Gene  \\\n",
       "7421   MTTSQKHRDFVAEPMGEKPVGSLAGIGEVLGKKLEERGFDKAYVVL...    O75531    BAF   \n",
       "14325  MLRAAARFGPRLGRRLLSAAATQAVPAPNQQPEVFCNQIFINNEWH...    P05091  ALDH2   \n",
       "20606  MAALGTVLFTGVRRLHCSVAAWAGGQWRLQQGLAANPSGYGPLTEL...    Q86TS9   RM52   \n",
       "\n",
       "                                          Identification  Protein_Length  \n",
       "7421   HUMAN Barrier-to-autointegration factor OS=Hom...              89  \n",
       "14325  HUMAN Aldehyde dehydrogenase, mitochondrial OS...             517  \n",
       "20606  HUMAN 39S ribosomal protein L52, mitochondrial...             123  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lists=[\"ALDH2\",\"BAF\",\"RM52\"]\n",
    "menu.loc[menu['Gene'].isin(lists)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Scales:\n",
    "    def Mass(peptide):\n",
    "        mass = {  \"A\": 71.037114,\n",
    "        \"R\": 156.101111,\n",
    "        \"N\": 114.042927,\n",
    "        \"D\": 115.026943,\n",
    "        \"C\": 103.009185,\n",
    "        \"Q\": 129.042593,\n",
    "        \"E\": 128.058578,\n",
    "        \"G\": 57.021464,\n",
    "        \"H\": 137.058912,\n",
    "        \"I\": 113.084064,\n",
    "        \"L\": 113.084064,\n",
    "        \"K\": 128.094963,\n",
    "        \"M\": 131.040485,\n",
    "        \"F\": 147.068414,\n",
    "        \"P\": 97.052764,\n",
    "        \"S\": 87.032028,\n",
    "        \"T\": 101.047679,\n",
    "        \"W\": 186.079313,\n",
    "        \"Y\": 163.06332,\n",
    "        \"V\": 99.068414,\n",
    "        }\n",
    "        mass_list = [mass.get(aa,0.0)for aa in peptide]\n",
    "        pep_mass=sum(mass_list)\n",
    "        return pep_mass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Protein(object):\n",
    "    ''' Orginal Protein Information (i.e. Mass, Length, etc.). \n",
    "    Primary use is to compare to peptides following digestion.\n",
    "    =============================================================\n",
    "    Hyperparameters\n",
    "    ---------------\n",
    "    Peptide = A str(peptide sequence)\n",
    "    Length = Automatically caluclated based on length of Peptide.\n",
    "    Mass = Automatically caluclated based on mass of Peptide.\n",
    "    --------------------------------------------------------------\n",
    "    ''' \n",
    "    def __init__(self,df,target,identifier):\n",
    "        self.protein_peptide=df[[target, identifier]].set_index(identifier).to_dict()[target]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "Human=Protein(menu,\"Peptide\",\"Gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MLVRRGARAGPRMPRGWTALCLLSLLPSGFMSLDNNGTATPELPTQGTFSNVSTNVSYQETTTPSTLGSTSLHPVSQHGNEATTNITETTVKFTSTSVITSVYGNTNSSVQSQTSVISTVFTTPANVSTPETTLKPSLSPGNVSDLSTTSTSLATSPTKPYTSSSPILSDIKAEIKCSGIREVKLTQGICLEQNKTSSCAEFKKDRGEGLARVLCGEEQADADAGAQVCSLLLAQSEVRPQCLLLVLANRTEISSKLQLMKKHQSDLKKLGILDFTEQDVASHQSYSQKTLIALVTSGALLAVLGITGYFLMNRRSWSPTGERLGEDPYYTENGGGQGYSSGPGTSPEAQGKASVNRGAQENGTGQATSRNGHSARQHVVADTEL'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Human.protein_peptide[\"CD34\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Menu(df,identity,sequence,db_id):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<__main__.Protein at 0x2799d1be9a0>,\n",
       " <__main__.Protein at 0x2799d1bea60>,\n",
       " <__main__.Protein at 0x2799d1beaf0>,\n",
       " <__main__.Protein at 0x2799d1beb50>,\n",
       " <__main__.Protein at 0x2799d1bebb0>,\n",
       " <__main__.Protein at 0x2799d1bec40>,\n",
       " <__main__.Protein at 0x2799d1beca0>,\n",
       " <__main__.Protein at 0x2799d1bed30>,\n",
       " <__main__.Protein at 0x2799d1bed90>,\n",
       " <__main__.Protein at 0x2799d1bedf0>,\n",
       " <__main__.Protein at 0x2799d1bee20>,\n",
       " <__main__.Protein at 0x2799d1bee80>,\n",
       " <__main__.Protein at 0x2799d1beee0>,\n",
       " <__main__.Protein at 0x2799d1bef10>,\n",
       " <__main__.Protein at 0x2799d1bef70>,\n",
       " <__main__.Protein at 0x2799d1befd0>,\n",
       " <__main__.Protein at 0x2799d1be040>,\n",
       " <__main__.Protein at 0x2799d1be550>,\n",
       " <__main__.Protein at 0x2799d1be670>,\n",
       " <__main__.Protein at 0x2799cfee040>,\n",
       " <__main__.Protein at 0x2799cfee0a0>,\n",
       " <__main__.Protein at 0x2799cfee100>,\n",
       " <__main__.Protein at 0x2799cfee160>,\n",
       " <__main__.Protein at 0x2799cfee1c0>,\n",
       " <__main__.Protein at 0x2799cfee250>,\n",
       " <__main__.Protein at 0x2799cfee2b0>,\n",
       " <__main__.Protein at 0x2799cfee310>,\n",
       " <__main__.Protein at 0x2799cfee370>,\n",
       " <__main__.Protein at 0x2799cfee3d0>,\n",
       " <__main__.Protein at 0x2799cfee430>,\n",
       " <__main__.Protein at 0x2799cfee490>,\n",
       " <__main__.Protein at 0x2799cfee4f0>,\n",
       " <__main__.Protein at 0x2799cfee550>,\n",
       " <__main__.Protein at 0x2799cfee5b0>,\n",
       " <__main__.Protein at 0x2799cfee610>,\n",
       " <__main__.Protein at 0x2799cfee670>,\n",
       " <__main__.Protein at 0x2799cfee6d0>,\n",
       " <__main__.Protein at 0x2799cfee730>,\n",
       " <__main__.Protein at 0x2799cfee790>,\n",
       " <__main__.Protein at 0x2799cfee7f0>,\n",
       " <__main__.Protein at 0x2799cfee850>,\n",
       " <__main__.Protein at 0x2799cfee8b0>,\n",
       " <__main__.Protein at 0x2799cfee910>,\n",
       " <__main__.Protein at 0x2799cfee970>,\n",
       " <__main__.Protein at 0x2799cfee9d0>,\n",
       " <__main__.Protein at 0x2799cfeea30>,\n",
       " <__main__.Protein at 0x2799cfeea90>,\n",
       " <__main__.Protein at 0x2799cfeeaf0>,\n",
       " <__main__.Protein at 0x2799cfeeb50>,\n",
       " <__main__.Protein at 0x2799cfeebb0>,\n",
       " <__main__.Protein at 0x2799cfeec10>,\n",
       " <__main__.Protein at 0x2799cfeec70>,\n",
       " <__main__.Protein at 0x2799cfeecd0>,\n",
       " <__main__.Protein at 0x2799cfeed30>,\n",
       " <__main__.Protein at 0x2799cfeed90>,\n",
       " <__main__.Protein at 0x2799cfeedf0>,\n",
       " <__main__.Protein at 0x2799cfeee50>,\n",
       " <__main__.Protein at 0x2799cfeeeb0>,\n",
       " <__main__.Protein at 0x2799cfeef10>,\n",
       " <__main__.Protein at 0x2799cfeef70>,\n",
       " <__main__.Protein at 0x2799cfeefd0>,\n",
       " <__main__.Protein at 0x27995965eb0>,\n",
       " <__main__.Protein at 0x279956421c0>,\n",
       " <__main__.Protein at 0x27995642370>,\n",
       " <__main__.Protein at 0x279956421f0>,\n",
       " <__main__.Protein at 0x27995642400>,\n",
       " <__main__.Protein at 0x27995642460>,\n",
       " <__main__.Protein at 0x27995642a60>,\n",
       " <__main__.Protein at 0x279956423d0>,\n",
       " <__main__.Protein at 0x2799b472ee0>,\n",
       " <__main__.Protein at 0x27995606af0>,\n",
       " <__main__.Protein at 0x27995606df0>,\n",
       " <__main__.Protein at 0x27997ee9100>,\n",
       " <__main__.Protein at 0x27995600cd0>,\n",
       " <__main__.Protein at 0x27995600460>,\n",
       " <__main__.Protein at 0x27995600520>,\n",
       " <__main__.Protein at 0x27995600400>,\n",
       " <__main__.Protein at 0x2799b461280>,\n",
       " <__main__.Protein at 0x2799dfd8f10>,\n",
       " <__main__.Protein at 0x2799dfd8af0>,\n",
       " <__main__.Protein at 0x2799dfd8a60>,\n",
       " <__main__.Protein at 0x2799dfd8dc0>,\n",
       " <__main__.Protein at 0x2799dfd8310>,\n",
       " <__main__.Protein at 0x2799dfd8970>,\n",
       " <__main__.Protein at 0x2799dfd8160>,\n",
       " <__main__.Protein at 0x2799dfd8cd0>,\n",
       " <__main__.Protein at 0x2799dfd8100>,\n",
       " <__main__.Protein at 0x2799dfd8f70>,\n",
       " <__main__.Protein at 0x2799dfd8130>,\n",
       " <__main__.Protein at 0x2799dfd8580>,\n",
       " <__main__.Protein at 0x2799dfd8850>,\n",
       " <__main__.Protein at 0x2799dfd8730>,\n",
       " <__main__.Protein at 0x2799dfd8490>,\n",
       " <__main__.Protein at 0x2799dfd84f0>,\n",
       " <__main__.Protein at 0x2799d1be2b0>,\n",
       " <__main__.Protein at 0x2799d1be3a0>,\n",
       " <__main__.Protein at 0x2799dfd8550>,\n",
       " <__main__.Protein at 0x2799dfd83d0>,\n",
       " <__main__.Protein at 0x2799dfd8520>,\n",
       " <__main__.Protein at 0x2799dfd8d00>,\n",
       " <__main__.Protein at 0x2799dfd89d0>,\n",
       " <__main__.Protein at 0x2799dfd8ee0>,\n",
       " <__main__.Protein at 0x2799dfd8040>,\n",
       " <__main__.Protein at 0x2799dfd8760>,\n",
       " <__main__.Protein at 0x2799dfd8070>,\n",
       " <__main__.Protein at 0x2799dfd8370>,\n",
       " <__main__.Protein at 0x2799dfd8c40>,\n",
       " <__main__.Protein at 0x2799dfd8d30>,\n",
       " <__main__.Protein at 0x2799dfd8e20>,\n",
       " <__main__.Protein at 0x2799dfd8940>,\n",
       " <__main__.Protein at 0x2799dfd88b0>,\n",
       " <__main__.Protein at 0x2799dfd8c10>,\n",
       " <__main__.Protein at 0x2799dfd83a0>,\n",
       " <__main__.Protein at 0x2799dfd8880>,\n",
       " <__main__.Protein at 0x2799dfd8700>,\n",
       " <__main__.Protein at 0x2799dfd81f0>,\n",
       " <__main__.Protein at 0x2799dfd8fa0>,\n",
       " <__main__.Protein at 0x2799dfd8fd0>,\n",
       " <__main__.Protein at 0x2799dfd8b80>,\n",
       " <__main__.Protein at 0x2799bda6b50>,\n",
       " <__main__.Protein at 0x2799bda6610>,\n",
       " <__main__.Protein at 0x2799bda6dc0>,\n",
       " <__main__.Protein at 0x2799bda67c0>,\n",
       " <__main__.Protein at 0x2799bda6d00>,\n",
       " <__main__.Protein at 0x2799bda6b20>,\n",
       " <__main__.Protein at 0x2799bda6970>,\n",
       " <__main__.Protein at 0x2799bda6a30>,\n",
       " <__main__.Protein at 0x2799bda6d60>,\n",
       " <__main__.Protein at 0x2799bda6e50>,\n",
       " <__main__.Protein at 0x2799bda6df0>,\n",
       " <__main__.Protein at 0x2799bda6820>,\n",
       " <__main__.Protein at 0x2799bda6e80>,\n",
       " <__main__.Protein at 0x2799bda6d30>,\n",
       " <__main__.Protein at 0x2799bda6190>,\n",
       " <__main__.Protein at 0x2799bda6d90>,\n",
       " <__main__.Protein at 0x2799bda6cd0>,\n",
       " <__main__.Protein at 0x2799bda6c70>,\n",
       " <__main__.Protein at 0x2799bda6c40>,\n",
       " <__main__.Protein at 0x2799bda6910>,\n",
       " <__main__.Protein at 0x2799bda6a90>,\n",
       " <__main__.Protein at 0x2799593b5e0>,\n",
       " <__main__.Protein at 0x2799dea5bb0>,\n",
       " <__main__.Protein at 0x2799dea5c70>,\n",
       " <__main__.Protein at 0x2799dea5ac0>,\n",
       " <__main__.Protein at 0x2799dea5ee0>,\n",
       " <__main__.Protein at 0x2799dea5610>,\n",
       " <__main__.Protein at 0x2799dea5730>,\n",
       " <__main__.Protein at 0x2799dea51f0>,\n",
       " <__main__.Protein at 0x2799dea5190>,\n",
       " <__main__.Protein at 0x2799dea5a60>,\n",
       " <__main__.Protein at 0x2799dea5af0>,\n",
       " <__main__.Protein at 0x2799dea5f40>,\n",
       " <__main__.Protein at 0x2799dea5f10>,\n",
       " <__main__.Protein at 0x2799dea5520>,\n",
       " <__main__.Protein at 0x2799dea54c0>,\n",
       " <__main__.Protein at 0x2799dea56d0>,\n",
       " <__main__.Protein at 0x2799dea5fa0>,\n",
       " <__main__.Protein at 0x2799dea5310>,\n",
       " <__main__.Protein at 0x2799dea5e20>,\n",
       " <__main__.Protein at 0x2799dea5250>,\n",
       " <__main__.Protein at 0x2799dea5b50>,\n",
       " <__main__.Protein at 0x2799dea57f0>,\n",
       " <__main__.Protein at 0x2799dea5640>,\n",
       " <__main__.Protein at 0x2799dea5370>,\n",
       " <__main__.Protein at 0x2799dea5ca0>,\n",
       " <__main__.Protein at 0x2799dea55e0>,\n",
       " <__main__.Protein at 0x2799dea5910>,\n",
       " <__main__.Protein at 0x2799dea5e80>,\n",
       " <__main__.Protein at 0x2799dea5460>,\n",
       " <__main__.Protein at 0x2799dea50a0>,\n",
       " <__main__.Protein at 0x2799dea59a0>,\n",
       " <__main__.Protein at 0x2799dea5100>,\n",
       " <__main__.Protein at 0x2799dea53d0>,\n",
       " <__main__.Protein at 0x2799dea5880>,\n",
       " <__main__.Protein at 0x2799de7d910>,\n",
       " <__main__.Protein at 0x2799de7d490>,\n",
       " <__main__.Protein at 0x2799de7d220>,\n",
       " <__main__.Protein at 0x2799de7d5b0>,\n",
       " <__main__.Protein at 0x2799de7d3d0>,\n",
       " <__main__.Protein at 0x2799de7d280>,\n",
       " <__main__.Protein at 0x2799de7d1f0>,\n",
       " <__main__.Protein at 0x2799de7dbb0>,\n",
       " <__main__.Protein at 0x2799de7d970>,\n",
       " <__main__.Protein at 0x2799de7dac0>,\n",
       " <__main__.Protein at 0x2799de7d3a0>,\n",
       " <__main__.Protein at 0x2799de7da90>,\n",
       " <__main__.Protein at 0x2799de7d4f0>,\n",
       " <__main__.Protein at 0x2799de7dbe0>,\n",
       " <__main__.Protein at 0x2799de7d550>,\n",
       " <__main__.Protein at 0x2799de7daf0>,\n",
       " <__main__.Protein at 0x2799de7db20>,\n",
       " <__main__.Protein at 0x2799de7d700>,\n",
       " <__main__.Protein at 0x2799de7dca0>,\n",
       " <__main__.Protein at 0x2799de7db80>,\n",
       " <__main__.Protein at 0x2799de7db50>,\n",
       " <__main__.Protein at 0x2799de7d9a0>,\n",
       " <__main__.Protein at 0x2799de7dd60>,\n",
       " <__main__.Protein at 0x2799de7ddc0>,\n",
       " <__main__.Protein at 0x2799de7de20>,\n",
       " <__main__.Protein at 0x2799de7de80>,\n",
       " <__main__.Protein at 0x2799de7dee0>,\n",
       " <__main__.Protein at 0x2799de7df40>,\n",
       " <__main__.Protein at 0x2799de7dfa0>,\n",
       " <__main__.Protein at 0x2799de7da00>,\n",
       " <__main__.Protein at 0x2799de7d850>,\n",
       " <__main__.Protein at 0x2799de7d340>,\n",
       " <__main__.Protein at 0x2799de7d610>,\n",
       " <__main__.Protein at 0x2799de7d190>,\n",
       " <__main__.Protein at 0x2799de7d160>,\n",
       " <__main__.Protein at 0x2799de7d100>,\n",
       " <__main__.Protein at 0x2799de7d130>,\n",
       " <__main__.Protein at 0x2799de7d790>,\n",
       " <__main__.Protein at 0x2799de7d7c0>,\n",
       " <__main__.Protein at 0x2799de7d250>,\n",
       " <__main__.Protein at 0x279955f5790>,\n",
       " <__main__.Protein at 0x279955f58b0>,\n",
       " <__main__.Protein at 0x279955f5d60>,\n",
       " <__main__.Protein at 0x279955f5cd0>,\n",
       " <__main__.Protein at 0x279955f5e50>,\n",
       " <__main__.Protein at 0x2799563b610>,\n",
       " <__main__.Protein at 0x2799563b670>,\n",
       " <__main__.Protein at 0x2799563bca0>,\n",
       " <__main__.Protein at 0x2799563bfa0>,\n",
       " <__main__.Protein at 0x2799563b370>,\n",
       " <__main__.Protein at 0x2799563b580>,\n",
       " <__main__.Protein at 0x2799563b5e0>,\n",
       " <__main__.Protein at 0x2799563b100>,\n",
       " <__main__.Protein at 0x2799563ba30>,\n",
       " <__main__.Protein at 0x2799563bbb0>,\n",
       " <__main__.Protein at 0x2799563b0a0>,\n",
       " <__main__.Protein at 0x2799563b6d0>,\n",
       " <__main__.Protein at 0x2799563b970>,\n",
       " <__main__.Protein at 0x2799563bb50>,\n",
       " <__main__.Protein at 0x2799563b880>,\n",
       " <__main__.Protein at 0x2799563b4c0>,\n",
       " <__main__.Protein at 0x2799563bd60>,\n",
       " <__main__.Protein at 0x2799563b4f0>,\n",
       " <__main__.Protein at 0x2799563baf0>,\n",
       " <__main__.Protein at 0x2799563bd90>,\n",
       " <__main__.Protein at 0x2799563b490>,\n",
       " <__main__.Protein at 0x2799563b8b0>,\n",
       " <__main__.Protein at 0x2799563bdf0>,\n",
       " <__main__.Protein at 0x2799563bc10>,\n",
       " <__main__.Protein at 0x2799563b430>,\n",
       " <__main__.Protein at 0x2799563b6a0>,\n",
       " <__main__.Protein at 0x2799563bd30>,\n",
       " <__main__.Protein at 0x2799563b250>,\n",
       " <__main__.Protein at 0x2799563bf40>,\n",
       " <__main__.Protein at 0x2799563ba00>,\n",
       " <__main__.Protein at 0x2799560cf70>,\n",
       " <__main__.Protein at 0x2799560cd30>,\n",
       " <__main__.Protein at 0x2799560c730>,\n",
       " <__main__.Protein at 0x2799560c100>,\n",
       " <__main__.Protein at 0x2799bdca070>,\n",
       " <__main__.Protein at 0x2799bdcaa30>,\n",
       " <__main__.Protein at 0x2799bdcab20>,\n",
       " <__main__.Protein at 0x2799bdca2b0>,\n",
       " <__main__.Protein at 0x2799bdca9a0>,\n",
       " <__main__.Protein at 0x2799bdca190>,\n",
       " <__main__.Protein at 0x2799bdca970>,\n",
       " <__main__.Protein at 0x2799bdca910>,\n",
       " <__main__.Protein at 0x2799bdca370>,\n",
       " <__main__.Protein at 0x2799be08b80>,\n",
       " <__main__.Protein at 0x2799be08ca0>,\n",
       " <__main__.Protein at 0x2799be08b20>,\n",
       " <__main__.Protein at 0x2799be089d0>,\n",
       " <__main__.Protein at 0x2799be08c70>,\n",
       " <__main__.Protein at 0x2799be08df0>,\n",
       " <__main__.Protein at 0x2799be08c40>,\n",
       " <__main__.Protein at 0x2799be08970>,\n",
       " <__main__.Protein at 0x2799be08bb0>,\n",
       " <__main__.Protein at 0x2799be08a90>,\n",
       " <__main__.Protein at 0x2799be08d30>,\n",
       " <__main__.Protein at 0x2799be08640>,\n",
       " <__main__.Protein at 0x2799be08f40>,\n",
       " <__main__.Protein at 0x2799be08760>,\n",
       " <__main__.Protein at 0x2799be08a30>,\n",
       " <__main__.Protein at 0x2799be08b50>,\n",
       " <__main__.Protein at 0x2799be085b0>,\n",
       " <__main__.Protein at 0x2799dccafa0>,\n",
       " <__main__.Protein at 0x2799dccad60>,\n",
       " <__main__.Protein at 0x2799dccaee0>,\n",
       " <__main__.Protein at 0x2799dccae20>,\n",
       " <__main__.Protein at 0x2799dccaf10>,\n",
       " <__main__.Protein at 0x2799dccae80>,\n",
       " <__main__.Protein at 0x2799dccafd0>,\n",
       " <__main__.Protein at 0x2799dccacd0>,\n",
       " <__main__.Protein at 0x2799dccabe0>,\n",
       " <__main__.Protein at 0x2799dcca040>,\n",
       " <__main__.Protein at 0x2799dcca0a0>,\n",
       " <__main__.Protein at 0x2799dcca100>,\n",
       " <__main__.Protein at 0x2799dcca160>,\n",
       " <__main__.Protein at 0x2799dcca1c0>,\n",
       " <__main__.Protein at 0x2799dcca220>,\n",
       " <__main__.Protein at 0x2799dcca280>,\n",
       " <__main__.Protein at 0x2799dcca2e0>,\n",
       " <__main__.Protein at 0x2799dcca340>,\n",
       " <__main__.Protein at 0x2799dcca3a0>,\n",
       " <__main__.Protein at 0x2799dcca400>,\n",
       " <__main__.Protein at 0x2799dcca460>,\n",
       " <__main__.Protein at 0x2799dcca4c0>,\n",
       " <__main__.Protein at 0x2799dcca520>,\n",
       " <__main__.Protein at 0x2799dcca580>,\n",
       " <__main__.Protein at 0x2799dcca5e0>,\n",
       " <__main__.Protein at 0x2799dcca640>,\n",
       " <__main__.Protein at 0x2799dcca6a0>,\n",
       " <__main__.Protein at 0x2799dcca700>,\n",
       " <__main__.Protein at 0x2799dcca760>,\n",
       " <__main__.Protein at 0x2799dcca7c0>,\n",
       " <__main__.Protein at 0x2799dcca820>,\n",
       " <__main__.Protein at 0x2799dcca880>,\n",
       " <__main__.Protein at 0x2799dcca8e0>,\n",
       " <__main__.Protein at 0x2799dcca940>,\n",
       " <__main__.Protein at 0x2799dcca9a0>,\n",
       " <__main__.Protein at 0x2799dccaa00>,\n",
       " <__main__.Protein at 0x2799dccaa60>,\n",
       " <__main__.Protein at 0x2799dac30a0>,\n",
       " <__main__.Protein at 0x2799dac3040>,\n",
       " <__main__.Protein at 0x2799dac3100>,\n",
       " <__main__.Protein at 0x2799dac3160>,\n",
       " <__main__.Protein at 0x2799e2cdbb0>,\n",
       " <__main__.Protein at 0x2799e2cdc10>,\n",
       " <__main__.Protein at 0x2799e2cdc70>,\n",
       " <__main__.Protein at 0x2799e2cdcd0>,\n",
       " <__main__.Protein at 0x2799e2cdd30>,\n",
       " <__main__.Protein at 0x2799e2cdd90>,\n",
       " <__main__.Protein at 0x2799e2cddf0>,\n",
       " <__main__.Protein at 0x2799e2cde50>,\n",
       " <__main__.Protein at 0x2799e2cdeb0>,\n",
       " <__main__.Protein at 0x2799e2cdf10>,\n",
       " <__main__.Protein at 0x2799e2cdf70>,\n",
       " <__main__.Protein at 0x2799e2cdfd0>,\n",
       " <__main__.Protein at 0x2799dae6070>,\n",
       " <__main__.Protein at 0x2799dae60d0>,\n",
       " <__main__.Protein at 0x2799dae6130>,\n",
       " <__main__.Protein at 0x2799dea5160>,\n",
       " <__main__.Protein at 0x2799dae61c0>,\n",
       " <__main__.Protein at 0x2799dae6220>,\n",
       " <__main__.Protein at 0x2799dae6280>,\n",
       " <__main__.Protein at 0x2799dae62e0>,\n",
       " <__main__.Protein at 0x2799dae6340>,\n",
       " <__main__.Protein at 0x2799dae63a0>,\n",
       " <__main__.Protein at 0x2799dae6400>,\n",
       " <__main__.Protein at 0x2799dae6460>,\n",
       " <__main__.Protein at 0x2799dae64c0>,\n",
       " <__main__.Protein at 0x2799dae6520>,\n",
       " <__main__.Protein at 0x2799dae6580>,\n",
       " <__main__.Protein at 0x2799dae65e0>,\n",
       " <__main__.Protein at 0x2799dae6640>,\n",
       " <__main__.Protein at 0x2799dae66a0>,\n",
       " <__main__.Protein at 0x2799dae6700>,\n",
       " <__main__.Protein at 0x2799dae6760>,\n",
       " <__main__.Protein at 0x2799dae67c0>,\n",
       " <__main__.Protein at 0x2799dae6820>,\n",
       " <__main__.Protein at 0x2799dae6880>,\n",
       " <__main__.Protein at 0x2799dae68e0>,\n",
       " <__main__.Protein at 0x2799dae6940>,\n",
       " <__main__.Protein at 0x2799dae69a0>,\n",
       " <__main__.Protein at 0x2799dae6a00>,\n",
       " <__main__.Protein at 0x2799dae6a60>,\n",
       " <__main__.Protein at 0x2799dae6ac0>,\n",
       " <__main__.Protein at 0x2799dae6b20>,\n",
       " <__main__.Protein at 0x2799dae6b80>,\n",
       " <__main__.Protein at 0x2799dae6be0>,\n",
       " <__main__.Protein at 0x2799dae6c40>,\n",
       " <__main__.Protein at 0x2799dae6ca0>,\n",
       " <__main__.Protein at 0x2799dae6d00>,\n",
       " <__main__.Protein at 0x2799dae6d60>,\n",
       " <__main__.Protein at 0x2799dae6dc0>,\n",
       " <__main__.Protein at 0x2799dae6e20>,\n",
       " <__main__.Protein at 0x2799dae6e80>,\n",
       " <__main__.Protein at 0x2799dae6ee0>,\n",
       " <__main__.Protein at 0x2799dae6f40>,\n",
       " <__main__.Protein at 0x2799dae6fa0>,\n",
       " <__main__.Protein at 0x2799be35040>,\n",
       " <__main__.Protein at 0x2799be350a0>,\n",
       " <__main__.Protein at 0x2799be35100>,\n",
       " <__main__.Protein at 0x2799be35160>,\n",
       " <__main__.Protein at 0x2799be351c0>,\n",
       " <__main__.Protein at 0x2799be35220>,\n",
       " <__main__.Protein at 0x2799be35280>,\n",
       " <__main__.Protein at 0x2799be352e0>,\n",
       " <__main__.Protein at 0x2799be35340>,\n",
       " <__main__.Protein at 0x2799be353a0>,\n",
       " <__main__.Protein at 0x2799be35400>,\n",
       " <__main__.Protein at 0x2799be35460>,\n",
       " <__main__.Protein at 0x2799be354c0>,\n",
       " <__main__.Protein at 0x2799be35520>,\n",
       " <__main__.Protein at 0x2799be35580>,\n",
       " <__main__.Protein at 0x2799be355e0>,\n",
       " <__main__.Protein at 0x2799be35640>,\n",
       " <__main__.Protein at 0x2799be356a0>,\n",
       " <__main__.Protein at 0x2799be35700>,\n",
       " <__main__.Protein at 0x2799be35760>,\n",
       " <__main__.Protein at 0x2799be357c0>,\n",
       " <__main__.Protein at 0x2799be35820>,\n",
       " <__main__.Protein at 0x2799be35880>,\n",
       " <__main__.Protein at 0x2799be358e0>,\n",
       " <__main__.Protein at 0x2799be35940>,\n",
       " <__main__.Protein at 0x2799be359a0>,\n",
       " <__main__.Protein at 0x2799be35a00>,\n",
       " <__main__.Protein at 0x2799be35a60>,\n",
       " <__main__.Protein at 0x2799be35ac0>,\n",
       " <__main__.Protein at 0x2799be35b20>,\n",
       " <__main__.Protein at 0x2799be35b80>,\n",
       " <__main__.Protein at 0x2799be35be0>,\n",
       " <__main__.Protein at 0x2799be35c40>,\n",
       " <__main__.Protein at 0x2799be35ca0>,\n",
       " <__main__.Protein at 0x2799be35d00>,\n",
       " <__main__.Protein at 0x2799be35d60>,\n",
       " <__main__.Protein at 0x2799be35dc0>,\n",
       " <__main__.Protein at 0x2799be35e20>,\n",
       " <__main__.Protein at 0x2799be35e80>,\n",
       " <__main__.Protein at 0x2799be35ee0>,\n",
       " <__main__.Protein at 0x2799be35f40>,\n",
       " <__main__.Protein at 0x2799be35fa0>,\n",
       " <__main__.Protein at 0x2799ddcc040>,\n",
       " <__main__.Protein at 0x2799ddcc0a0>,\n",
       " <__main__.Protein at 0x2799ddcc100>,\n",
       " <__main__.Protein at 0x2799ddcc160>,\n",
       " <__main__.Protein at 0x2799ddcc1c0>,\n",
       " <__main__.Protein at 0x2799ddcc220>,\n",
       " <__main__.Protein at 0x2799ddcc280>,\n",
       " <__main__.Protein at 0x2799ddcc2e0>,\n",
       " <__main__.Protein at 0x2799ddcc340>,\n",
       " <__main__.Protein at 0x2799ddcc3a0>,\n",
       " <__main__.Protein at 0x2799ddcc400>,\n",
       " <__main__.Protein at 0x2799ddcc460>,\n",
       " <__main__.Protein at 0x2799ddcc4c0>,\n",
       " <__main__.Protein at 0x2799ddcc520>,\n",
       " <__main__.Protein at 0x2799ddcc580>,\n",
       " <__main__.Protein at 0x2799ddcc5e0>,\n",
       " <__main__.Protein at 0x2799ddcc640>,\n",
       " <__main__.Protein at 0x2799ddcc6a0>,\n",
       " <__main__.Protein at 0x2799ddcc700>,\n",
       " <__main__.Protein at 0x2799ddcc760>,\n",
       " <__main__.Protein at 0x2799ddcc7c0>,\n",
       " <__main__.Protein at 0x2799ddcc820>,\n",
       " <__main__.Protein at 0x2799ddcc880>,\n",
       " <__main__.Protein at 0x2799ddcc8e0>,\n",
       " <__main__.Protein at 0x2799ddcc940>,\n",
       " <__main__.Protein at 0x2799ddcc9a0>,\n",
       " <__main__.Protein at 0x2799ddcca00>,\n",
       " <__main__.Protein at 0x2799ddcca60>,\n",
       " <__main__.Protein at 0x2799ddccac0>,\n",
       " <__main__.Protein at 0x2799ddccb20>,\n",
       " <__main__.Protein at 0x2799ddccb80>,\n",
       " <__main__.Protein at 0x2799ddccbe0>,\n",
       " <__main__.Protein at 0x2799ddccc40>,\n",
       " <__main__.Protein at 0x2799ddccca0>,\n",
       " <__main__.Protein at 0x2799ddccd00>,\n",
       " <__main__.Protein at 0x2799ddccd60>,\n",
       " <__main__.Protein at 0x2799ddccdc0>,\n",
       " <__main__.Protein at 0x2799ddcce20>,\n",
       " <__main__.Protein at 0x2799ddcce80>,\n",
       " <__main__.Protein at 0x2799ddccee0>,\n",
       " <__main__.Protein at 0x2799ddccf40>,\n",
       " <__main__.Protein at 0x2799ddccfa0>,\n",
       " <__main__.Protein at 0x2799ddc2040>,\n",
       " <__main__.Protein at 0x2799ddc20a0>,\n",
       " <__main__.Protein at 0x2799ddc2100>,\n",
       " <__main__.Protein at 0x2799ddc2160>,\n",
       " <__main__.Protein at 0x2799ddc21c0>,\n",
       " <__main__.Protein at 0x2799ddc2220>,\n",
       " <__main__.Protein at 0x2799ddc2280>,\n",
       " <__main__.Protein at 0x2799ddc22e0>,\n",
       " <__main__.Protein at 0x2799ddc2340>,\n",
       " <__main__.Protein at 0x2799ddc23a0>,\n",
       " <__main__.Protein at 0x2799ddc2400>,\n",
       " <__main__.Protein at 0x2799ddc2460>,\n",
       " <__main__.Protein at 0x2799ddc24c0>,\n",
       " <__main__.Protein at 0x2799ddc2520>,\n",
       " <__main__.Protein at 0x2799ddc2580>,\n",
       " <__main__.Protein at 0x2799ddc25e0>,\n",
       " <__main__.Protein at 0x2799ddc2640>,\n",
       " <__main__.Protein at 0x2799ddc26a0>,\n",
       " <__main__.Protein at 0x2799ddc2700>,\n",
       " <__main__.Protein at 0x2799ddc2760>,\n",
       " <__main__.Protein at 0x2799ddc27c0>,\n",
       " <__main__.Protein at 0x2799ddc2820>,\n",
       " <__main__.Protein at 0x2799ddc2880>,\n",
       " <__main__.Protein at 0x2799ddc28e0>,\n",
       " <__main__.Protein at 0x2799ddc2940>,\n",
       " <__main__.Protein at 0x2799ddc29a0>,\n",
       " <__main__.Protein at 0x2799ddc2a00>,\n",
       " <__main__.Protein at 0x2799ddc2a60>,\n",
       " <__main__.Protein at 0x2799ddc2ac0>,\n",
       " <__main__.Protein at 0x2799ddc2b20>,\n",
       " <__main__.Protein at 0x2799ddc2b80>,\n",
       " <__main__.Protein at 0x2799ddc2be0>,\n",
       " <__main__.Protein at 0x2799ddc2c40>,\n",
       " <__main__.Protein at 0x2799ddc2ca0>,\n",
       " <__main__.Protein at 0x2799ddc2d00>,\n",
       " <__main__.Protein at 0x2799ddc2d60>,\n",
       " <__main__.Protein at 0x2799ddc2dc0>,\n",
       " <__main__.Protein at 0x2799ddc2e20>,\n",
       " <__main__.Protein at 0x2799ddc2e80>,\n",
       " <__main__.Protein at 0x2799ddc2ee0>,\n",
       " <__main__.Protein at 0x2799ddc2f40>,\n",
       " <__main__.Protein at 0x2799ddc2fa0>,\n",
       " <__main__.Protein at 0x2799ddc7040>,\n",
       " <__main__.Protein at 0x2799ddc70a0>,\n",
       " <__main__.Protein at 0x2799ddc7100>,\n",
       " <__main__.Protein at 0x2799ddc7160>,\n",
       " <__main__.Protein at 0x2799ddc71c0>,\n",
       " <__main__.Protein at 0x2799ddc7220>,\n",
       " <__main__.Protein at 0x2799ddc7280>,\n",
       " <__main__.Protein at 0x2799ddc72e0>,\n",
       " <__main__.Protein at 0x2799ddc7340>,\n",
       " <__main__.Protein at 0x2799ddc73a0>,\n",
       " <__main__.Protein at 0x2799ddc7400>,\n",
       " <__main__.Protein at 0x2799ddc7460>,\n",
       " <__main__.Protein at 0x2799ddc74c0>,\n",
       " <__main__.Protein at 0x2799ddc7520>,\n",
       " <__main__.Protein at 0x2799ddc7580>,\n",
       " <__main__.Protein at 0x2799ddc75e0>,\n",
       " <__main__.Protein at 0x2799ddc7640>,\n",
       " <__main__.Protein at 0x2799ddc76a0>,\n",
       " <__main__.Protein at 0x2799ddc7700>,\n",
       " <__main__.Protein at 0x2799ddc7760>,\n",
       " <__main__.Protein at 0x2799ddc77c0>,\n",
       " <__main__.Protein at 0x2799ddc7820>,\n",
       " <__main__.Protein at 0x2799ddc7880>,\n",
       " <__main__.Protein at 0x2799ddc78e0>,\n",
       " <__main__.Protein at 0x2799ddc7940>,\n",
       " <__main__.Protein at 0x2799ddc79a0>,\n",
       " <__main__.Protein at 0x2799ddc7a00>,\n",
       " <__main__.Protein at 0x2799ddc7a60>,\n",
       " <__main__.Protein at 0x2799ddc7ac0>,\n",
       " <__main__.Protein at 0x2799ddc7b20>,\n",
       " <__main__.Protein at 0x2799ddc7b80>,\n",
       " <__main__.Protein at 0x2799ddc7be0>,\n",
       " <__main__.Protein at 0x2799ddc7c40>,\n",
       " <__main__.Protein at 0x2799ddc7ca0>,\n",
       " <__main__.Protein at 0x2799ddc7d00>,\n",
       " <__main__.Protein at 0x2799ddc7d60>,\n",
       " <__main__.Protein at 0x2799ddc7dc0>,\n",
       " <__main__.Protein at 0x2799ddc7e20>,\n",
       " <__main__.Protein at 0x2799ddc7e80>,\n",
       " <__main__.Protein at 0x2799ddc7ee0>,\n",
       " <__main__.Protein at 0x2799ddc7f40>,\n",
       " <__main__.Protein at 0x2799ddc7fa0>,\n",
       " <__main__.Protein at 0x2799d66b040>,\n",
       " <__main__.Protein at 0x2799d66b0a0>,\n",
       " <__main__.Protein at 0x2799d66b100>,\n",
       " <__main__.Protein at 0x2799d66b160>,\n",
       " <__main__.Protein at 0x2799d66b1c0>,\n",
       " <__main__.Protein at 0x2799d66b220>,\n",
       " <__main__.Protein at 0x2799d66b280>,\n",
       " <__main__.Protein at 0x2799d66b2e0>,\n",
       " <__main__.Protein at 0x2799d66b340>,\n",
       " <__main__.Protein at 0x2799d66b3a0>,\n",
       " <__main__.Protein at 0x2799d66b400>,\n",
       " <__main__.Protein at 0x2799d66b460>,\n",
       " <__main__.Protein at 0x2799d66b4c0>,\n",
       " <__main__.Protein at 0x2799d66b520>,\n",
       " <__main__.Protein at 0x2799d66b580>,\n",
       " <__main__.Protein at 0x2799d66b5e0>,\n",
       " <__main__.Protein at 0x2799d66b640>,\n",
       " <__main__.Protein at 0x2799d66b6a0>,\n",
       " <__main__.Protein at 0x2799d66b700>,\n",
       " <__main__.Protein at 0x2799dac31f0>,\n",
       " <__main__.Protein at 0x2799d66b760>,\n",
       " <__main__.Protein at 0x2799d66b7c0>,\n",
       " <__main__.Protein at 0x2799d66b820>,\n",
       " <__main__.Protein at 0x2799d66b880>,\n",
       " <__main__.Protein at 0x2799d66b8e0>,\n",
       " <__main__.Protein at 0x2799d66b940>,\n",
       " <__main__.Protein at 0x2799d66b9a0>,\n",
       " <__main__.Protein at 0x2799d66ba00>,\n",
       " <__main__.Protein at 0x2799d66ba60>,\n",
       " <__main__.Protein at 0x2799d66bac0>,\n",
       " <__main__.Protein at 0x2799d66bb20>,\n",
       " <__main__.Protein at 0x2799d66bb80>,\n",
       " <__main__.Protein at 0x2799d66bbe0>,\n",
       " <__main__.Protein at 0x2799d66bc40>,\n",
       " <__main__.Protein at 0x2799d66bca0>,\n",
       " <__main__.Protein at 0x2799d66bd00>,\n",
       " <__main__.Protein at 0x2799d66bd60>,\n",
       " <__main__.Protein at 0x2799d66bdc0>,\n",
       " <__main__.Protein at 0x2799d66be20>,\n",
       " <__main__.Protein at 0x2799d66be80>,\n",
       " <__main__.Protein at 0x2799d66bee0>,\n",
       " <__main__.Protein at 0x2799d66bf40>,\n",
       " <__main__.Protein at 0x2799d66bfa0>,\n",
       " <__main__.Protein at 0x2799d679040>,\n",
       " <__main__.Protein at 0x2799d6790a0>,\n",
       " <__main__.Protein at 0x2799d679100>,\n",
       " <__main__.Protein at 0x2799d679160>,\n",
       " <__main__.Protein at 0x2799d6791c0>,\n",
       " <__main__.Protein at 0x2799d679220>,\n",
       " <__main__.Protein at 0x2799d679280>,\n",
       " <__main__.Protein at 0x2799d6792e0>,\n",
       " <__main__.Protein at 0x2799d679340>,\n",
       " <__main__.Protein at 0x2799d6793a0>,\n",
       " <__main__.Protein at 0x2799d679400>,\n",
       " <__main__.Protein at 0x2799d679460>,\n",
       " <__main__.Protein at 0x2799d6794c0>,\n",
       " <__main__.Protein at 0x2799d679520>,\n",
       " <__main__.Protein at 0x2799d679580>,\n",
       " <__main__.Protein at 0x2799d6795e0>,\n",
       " <__main__.Protein at 0x2799d679640>,\n",
       " <__main__.Protein at 0x2799d6796a0>,\n",
       " <__main__.Protein at 0x2799d679700>,\n",
       " <__main__.Protein at 0x2799d679760>,\n",
       " <__main__.Protein at 0x2799d6797c0>,\n",
       " <__main__.Protein at 0x2799d679820>,\n",
       " <__main__.Protein at 0x2799d679880>,\n",
       " <__main__.Protein at 0x2799d6798e0>,\n",
       " <__main__.Protein at 0x2799d679940>,\n",
       " <__main__.Protein at 0x2799d6799a0>,\n",
       " <__main__.Protein at 0x2799d679a00>,\n",
       " <__main__.Protein at 0x2799d679a60>,\n",
       " <__main__.Protein at 0x2799d679ac0>,\n",
       " <__main__.Protein at 0x2799d679b20>,\n",
       " <__main__.Protein at 0x2799d679b80>,\n",
       " <__main__.Protein at 0x2799d679be0>,\n",
       " <__main__.Protein at 0x2799d679c40>,\n",
       " <__main__.Protein at 0x2799d679ca0>,\n",
       " <__main__.Protein at 0x2799d679d00>,\n",
       " <__main__.Protein at 0x2799d679d60>,\n",
       " <__main__.Protein at 0x2799d679dc0>,\n",
       " <__main__.Protein at 0x2799d679e20>,\n",
       " <__main__.Protein at 0x2799d679e80>,\n",
       " <__main__.Protein at 0x2799d679ee0>,\n",
       " <__main__.Protein at 0x2799d679f40>,\n",
       " <__main__.Protein at 0x2799d679fa0>,\n",
       " <__main__.Protein at 0x2799d678040>,\n",
       " <__main__.Protein at 0x2799d6780a0>,\n",
       " <__main__.Protein at 0x2799d678100>,\n",
       " <__main__.Protein at 0x2799d678160>,\n",
       " <__main__.Protein at 0x2799d6781c0>,\n",
       " <__main__.Protein at 0x2799d678220>,\n",
       " <__main__.Protein at 0x2799d678280>,\n",
       " <__main__.Protein at 0x2799d6782e0>,\n",
       " <__main__.Protein at 0x2799d678340>,\n",
       " <__main__.Protein at 0x2799d6783a0>,\n",
       " <__main__.Protein at 0x2799d678400>,\n",
       " <__main__.Protein at 0x2799d678460>,\n",
       " <__main__.Protein at 0x2799d6784c0>,\n",
       " <__main__.Protein at 0x2799d678520>,\n",
       " <__main__.Protein at 0x2799d678580>,\n",
       " <__main__.Protein at 0x2799d6785e0>,\n",
       " <__main__.Protein at 0x2799d678640>,\n",
       " <__main__.Protein at 0x2799d6786a0>,\n",
       " <__main__.Protein at 0x2799d678700>,\n",
       " <__main__.Protein at 0x2799d678760>,\n",
       " <__main__.Protein at 0x2799d6787c0>,\n",
       " <__main__.Protein at 0x2799d678820>,\n",
       " <__main__.Protein at 0x2799d678880>,\n",
       " <__main__.Protein at 0x2799d6788e0>,\n",
       " <__main__.Protein at 0x2799d678940>,\n",
       " <__main__.Protein at 0x2799d6789a0>,\n",
       " <__main__.Protein at 0x2799d678a00>,\n",
       " <__main__.Protein at 0x2799d678a60>,\n",
       " <__main__.Protein at 0x2799d678ac0>,\n",
       " <__main__.Protein at 0x2799d678b20>,\n",
       " <__main__.Protein at 0x2799d678b80>,\n",
       " <__main__.Protein at 0x2799d678be0>,\n",
       " <__main__.Protein at 0x2799d678c40>,\n",
       " <__main__.Protein at 0x2799d678ca0>,\n",
       " <__main__.Protein at 0x2799d678d00>,\n",
       " <__main__.Protein at 0x2799d678d60>,\n",
       " <__main__.Protein at 0x2799d678dc0>,\n",
       " <__main__.Protein at 0x2799d678e20>,\n",
       " <__main__.Protein at 0x2799d678e80>,\n",
       " <__main__.Protein at 0x2799d678ee0>,\n",
       " <__main__.Protein at 0x2799d678f40>,\n",
       " <__main__.Protein at 0x2799d678fa0>,\n",
       " <__main__.Protein at 0x2799d66c040>,\n",
       " <__main__.Protein at 0x2799d66c0a0>,\n",
       " <__main__.Protein at 0x2799d66c100>,\n",
       " <__main__.Protein at 0x2799d66c160>,\n",
       " <__main__.Protein at 0x2799d66c1c0>,\n",
       " <__main__.Protein at 0x2799d66c220>,\n",
       " <__main__.Protein at 0x2799d66c280>,\n",
       " <__main__.Protein at 0x2799d66c2e0>,\n",
       " <__main__.Protein at 0x2799d66c340>,\n",
       " <__main__.Protein at 0x2799d8cea60>,\n",
       " <__main__.Protein at 0x2799d1be1c0>,\n",
       " <__main__.Protein at 0x2799d1be370>,\n",
       " <__main__.Protein at 0x2799d1be160>,\n",
       " <__main__.Protein at 0x2799d66c3d0>,\n",
       " <__main__.Protein at 0x2799d66c430>,\n",
       " <__main__.Protein at 0x2799d66c490>,\n",
       " <__main__.Protein at 0x2799d66c4f0>,\n",
       " <__main__.Protein at 0x2799d66c550>,\n",
       " <__main__.Protein at 0x2799d66c5b0>,\n",
       " <__main__.Protein at 0x2799d66c610>,\n",
       " <__main__.Protein at 0x2799d66c670>,\n",
       " <__main__.Protein at 0x2799d66c6d0>,\n",
       " <__main__.Protein at 0x2799d66c730>,\n",
       " <__main__.Protein at 0x2799d66c790>,\n",
       " <__main__.Protein at 0x2799d66c7f0>,\n",
       " <__main__.Protein at 0x2799d66c850>,\n",
       " <__main__.Protein at 0x2799d66c8b0>,\n",
       " <__main__.Protein at 0x2799d66c910>,\n",
       " <__main__.Protein at 0x2799d66c970>,\n",
       " <__main__.Protein at 0x2799d66c9d0>,\n",
       " <__main__.Protein at 0x2799d66ca30>,\n",
       " <__main__.Protein at 0x2799d66ca90>,\n",
       " <__main__.Protein at 0x2799d66caf0>,\n",
       " <__main__.Protein at 0x2799d66cb50>,\n",
       " <__main__.Protein at 0x2799d66cbb0>,\n",
       " <__main__.Protein at 0x2799d66cc10>,\n",
       " <__main__.Protein at 0x2799d66cc70>,\n",
       " <__main__.Protein at 0x2799d66ccd0>,\n",
       " <__main__.Protein at 0x2799d66cd30>,\n",
       " <__main__.Protein at 0x2799d66cd90>,\n",
       " <__main__.Protein at 0x2799d66cdf0>,\n",
       " <__main__.Protein at 0x2799d66ce50>,\n",
       " <__main__.Protein at 0x2799d66ceb0>,\n",
       " <__main__.Protein at 0x2799d66cf10>,\n",
       " <__main__.Protein at 0x2799d66cf70>,\n",
       " <__main__.Protein at 0x2799d66cfd0>,\n",
       " <__main__.Protein at 0x2799d665070>,\n",
       " <__main__.Protein at 0x2799d6650d0>,\n",
       " <__main__.Protein at 0x2799d665130>,\n",
       " <__main__.Protein at 0x2799d665190>,\n",
       " <__main__.Protein at 0x2799d6651f0>,\n",
       " <__main__.Protein at 0x2799d665250>,\n",
       " <__main__.Protein at 0x2799d6652b0>,\n",
       " <__main__.Protein at 0x2799d665310>,\n",
       " <__main__.Protein at 0x2799d665370>,\n",
       " <__main__.Protein at 0x2799d6653d0>,\n",
       " <__main__.Protein at 0x2799d665430>,\n",
       " <__main__.Protein at 0x2799d665490>,\n",
       " <__main__.Protein at 0x2799d6654f0>,\n",
       " <__main__.Protein at 0x2799d665550>,\n",
       " <__main__.Protein at 0x2799d6655b0>,\n",
       " <__main__.Protein at 0x2799d665610>,\n",
       " <__main__.Protein at 0x2799d665670>,\n",
       " <__main__.Protein at 0x2799d6656d0>,\n",
       " <__main__.Protein at 0x2799d665730>,\n",
       " <__main__.Protein at 0x2799d665790>,\n",
       " <__main__.Protein at 0x2799d6657f0>,\n",
       " <__main__.Protein at 0x2799d665850>,\n",
       " <__main__.Protein at 0x2799d6658b0>,\n",
       " <__main__.Protein at 0x2799d665910>,\n",
       " <__main__.Protein at 0x2799d665970>,\n",
       " <__main__.Protein at 0x2799d6659d0>,\n",
       " <__main__.Protein at 0x2799d665a30>,\n",
       " <__main__.Protein at 0x2799d665a90>,\n",
       " <__main__.Protein at 0x2799d665af0>,\n",
       " <__main__.Protein at 0x2799d665b50>,\n",
       " <__main__.Protein at 0x2799d665bb0>,\n",
       " <__main__.Protein at 0x2799d665c10>,\n",
       " <__main__.Protein at 0x2799d665c70>,\n",
       " <__main__.Protein at 0x2799d665cd0>,\n",
       " <__main__.Protein at 0x2799d665d30>,\n",
       " <__main__.Protein at 0x2799d665d90>,\n",
       " <__main__.Protein at 0x2799d665df0>,\n",
       " <__main__.Protein at 0x2799d665e50>,\n",
       " <__main__.Protein at 0x2799d665eb0>,\n",
       " <__main__.Protein at 0x2799d665f10>,\n",
       " <__main__.Protein at 0x2799d665f70>,\n",
       " <__main__.Protein at 0x2799d665fd0>,\n",
       " <__main__.Protein at 0x2799d666070>,\n",
       " <__main__.Protein at 0x2799d6660d0>,\n",
       " <__main__.Protein at 0x2799d666130>,\n",
       " <__main__.Protein at 0x2799d666190>,\n",
       " <__main__.Protein at 0x2799d6661f0>,\n",
       " <__main__.Protein at 0x2799d666250>,\n",
       " <__main__.Protein at 0x2799d6662b0>,\n",
       " <__main__.Protein at 0x2799d666310>,\n",
       " <__main__.Protein at 0x2799d666370>,\n",
       " <__main__.Protein at 0x2799d6663d0>,\n",
       " <__main__.Protein at 0x2799d666430>,\n",
       " <__main__.Protein at 0x2799d666490>,\n",
       " <__main__.Protein at 0x2799d6664f0>,\n",
       " <__main__.Protein at 0x2799d666550>,\n",
       " <__main__.Protein at 0x2799d6665b0>,\n",
       " <__main__.Protein at 0x2799d666610>,\n",
       " <__main__.Protein at 0x2799d666670>,\n",
       " <__main__.Protein at 0x2799d6666d0>,\n",
       " <__main__.Protein at 0x2799d666730>,\n",
       " <__main__.Protein at 0x2799d666790>,\n",
       " <__main__.Protein at 0x2799d6667f0>,\n",
       " <__main__.Protein at 0x2799d666850>,\n",
       " <__main__.Protein at 0x2799d6668b0>,\n",
       " <__main__.Protein at 0x2799d666910>,\n",
       " <__main__.Protein at 0x2799d666970>,\n",
       " <__main__.Protein at 0x2799d6669d0>,\n",
       " <__main__.Protein at 0x2799d666a30>,\n",
       " <__main__.Protein at 0x2799d666a90>,\n",
       " <__main__.Protein at 0x2799d666af0>,\n",
       " <__main__.Protein at 0x2799d666b50>,\n",
       " <__main__.Protein at 0x2799d666bb0>,\n",
       " <__main__.Protein at 0x2799d666c10>,\n",
       " <__main__.Protein at 0x2799d666c70>,\n",
       " <__main__.Protein at 0x2799d666cd0>,\n",
       " <__main__.Protein at 0x2799d666d30>,\n",
       " <__main__.Protein at 0x2799d666d90>,\n",
       " <__main__.Protein at 0x2799d666df0>,\n",
       " <__main__.Protein at 0x2799d666e50>,\n",
       " <__main__.Protein at 0x2799d666eb0>,\n",
       " <__main__.Protein at 0x2799d666f10>,\n",
       " <__main__.Protein at 0x2799d666f70>,\n",
       " <__main__.Protein at 0x2799d666fd0>,\n",
       " <__main__.Protein at 0x2799d654070>,\n",
       " <__main__.Protein at 0x2799d6540d0>,\n",
       " <__main__.Protein at 0x2799d654130>,\n",
       " <__main__.Protein at 0x2799d654190>,\n",
       " <__main__.Protein at 0x2799d6541f0>,\n",
       " <__main__.Protein at 0x2799d654250>,\n",
       " <__main__.Protein at 0x2799d6542b0>,\n",
       " <__main__.Protein at 0x2799d654310>,\n",
       " <__main__.Protein at 0x2799d654370>,\n",
       " <__main__.Protein at 0x2799d6543d0>,\n",
       " <__main__.Protein at 0x2799d654430>,\n",
       " <__main__.Protein at 0x2799d654490>,\n",
       " <__main__.Protein at 0x2799d6544f0>,\n",
       " <__main__.Protein at 0x2799d654550>,\n",
       " <__main__.Protein at 0x2799d6545b0>,\n",
       " <__main__.Protein at 0x2799d654610>,\n",
       " <__main__.Protein at 0x2799d654670>,\n",
       " <__main__.Protein at 0x2799d6546d0>,\n",
       " <__main__.Protein at 0x2799d654730>,\n",
       " <__main__.Protein at 0x2799d654790>,\n",
       " <__main__.Protein at 0x2799d6547f0>,\n",
       " <__main__.Protein at 0x2799d654850>,\n",
       " <__main__.Protein at 0x2799d6548b0>,\n",
       " <__main__.Protein at 0x2799d654910>,\n",
       " <__main__.Protein at 0x2799d654970>,\n",
       " <__main__.Protein at 0x2799d6549d0>,\n",
       " <__main__.Protein at 0x2799d654a30>,\n",
       " <__main__.Protein at 0x2799d654a90>,\n",
       " <__main__.Protein at 0x2799d654af0>,\n",
       " <__main__.Protein at 0x2799d654b50>,\n",
       " <__main__.Protein at 0x2799d654bb0>,\n",
       " <__main__.Protein at 0x2799d654c10>,\n",
       " <__main__.Protein at 0x2799d654c70>,\n",
       " <__main__.Protein at 0x2799d654cd0>,\n",
       " <__main__.Protein at 0x2799d654d30>,\n",
       " <__main__.Protein at 0x2799d654d90>,\n",
       " <__main__.Protein at 0x2799d654df0>,\n",
       " <__main__.Protein at 0x2799d654e50>,\n",
       " <__main__.Protein at 0x2799d654eb0>,\n",
       " <__main__.Protein at 0x2799d654f10>,\n",
       " <__main__.Protein at 0x2799d654f70>,\n",
       " <__main__.Protein at 0x2799d654fd0>,\n",
       " <__main__.Protein at 0x2799d648070>,\n",
       " <__main__.Protein at 0x2799d6480d0>,\n",
       " <__main__.Protein at 0x2799d648130>,\n",
       " <__main__.Protein at 0x2799d648190>,\n",
       " <__main__.Protein at 0x2799d6481f0>,\n",
       " <__main__.Protein at 0x2799d648250>,\n",
       " <__main__.Protein at 0x2799d6482b0>,\n",
       " <__main__.Protein at 0x2799d648310>,\n",
       " <__main__.Protein at 0x2799d648370>,\n",
       " <__main__.Protein at 0x2799d6483d0>,\n",
       " <__main__.Protein at 0x2799d648430>,\n",
       " <__main__.Protein at 0x2799d648490>,\n",
       " <__main__.Protein at 0x2799d6484f0>,\n",
       " <__main__.Protein at 0x2799d648550>,\n",
       " <__main__.Protein at 0x2799d6485b0>,\n",
       " <__main__.Protein at 0x2799d648610>,\n",
       " <__main__.Protein at 0x2799d648670>,\n",
       " <__main__.Protein at 0x2799d6486d0>,\n",
       " <__main__.Protein at 0x2799d648730>,\n",
       " <__main__.Protein at 0x2799d648790>,\n",
       " <__main__.Protein at 0x2799d6487f0>,\n",
       " <__main__.Protein at 0x2799d648850>,\n",
       " <__main__.Protein at 0x2799d6488b0>,\n",
       " <__main__.Protein at 0x2799d648910>,\n",
       " <__main__.Protein at 0x2799d648970>,\n",
       " <__main__.Protein at 0x2799d6489d0>,\n",
       " <__main__.Protein at 0x2799d648a30>,\n",
       " <__main__.Protein at 0x2799d648a90>,\n",
       " <__main__.Protein at 0x2799d648af0>,\n",
       " <__main__.Protein at 0x2799d648b50>,\n",
       " <__main__.Protein at 0x2799d648bb0>,\n",
       " <__main__.Protein at 0x2799d648c10>,\n",
       " <__main__.Protein at 0x2799d648c70>,\n",
       " <__main__.Protein at 0x2799d648cd0>,\n",
       " <__main__.Protein at 0x2799d648d30>,\n",
       " <__main__.Protein at 0x2799d648d90>,\n",
       " <__main__.Protein at 0x2799d648df0>,\n",
       " <__main__.Protein at 0x2799d648e50>,\n",
       " <__main__.Protein at 0x2799d648eb0>,\n",
       " <__main__.Protein at 0x2799d648f10>,\n",
       " <__main__.Protein at 0x2799d648f70>,\n",
       " <__main__.Protein at 0x2799d648fd0>,\n",
       " <__main__.Protein at 0x2799d2c5070>,\n",
       " <__main__.Protein at 0x2799d2c50d0>,\n",
       " <__main__.Protein at 0x2799d2c5130>,\n",
       " <__main__.Protein at 0x2799d2c5190>,\n",
       " <__main__.Protein at 0x2799d2c51f0>,\n",
       " <__main__.Protein at 0x2799d2c5250>,\n",
       " <__main__.Protein at 0x2799d2c52b0>,\n",
       " <__main__.Protein at 0x2799d2c5310>,\n",
       " <__main__.Protein at 0x2799d2c5370>,\n",
       " <__main__.Protein at 0x2799d2c53d0>,\n",
       " <__main__.Protein at 0x2799d2c5430>,\n",
       " <__main__.Protein at 0x2799d2c5490>,\n",
       " <__main__.Protein at 0x2799d2c54f0>,\n",
       " <__main__.Protein at 0x2799d2c5550>,\n",
       " <__main__.Protein at 0x2799d2c55b0>,\n",
       " <__main__.Protein at 0x2799d2c5610>,\n",
       " <__main__.Protein at 0x2799d2c5670>,\n",
       " <__main__.Protein at 0x2799d2c56d0>,\n",
       " <__main__.Protein at 0x2799d2c5730>,\n",
       " <__main__.Protein at 0x2799d2c5790>,\n",
       " <__main__.Protein at 0x2799d2c57f0>,\n",
       " <__main__.Protein at 0x2799d2c5850>,\n",
       " <__main__.Protein at 0x2799d2c58b0>,\n",
       " <__main__.Protein at 0x2799d2c5910>,\n",
       " <__main__.Protein at 0x2799d2c5970>,\n",
       " <__main__.Protein at 0x2799d2c59d0>,\n",
       " <__main__.Protein at 0x2799d2c5a30>,\n",
       " <__main__.Protein at 0x2799d2c5a90>,\n",
       " <__main__.Protein at 0x2799d2c5af0>,\n",
       " <__main__.Protein at 0x2799d2c5b50>,\n",
       " <__main__.Protein at 0x2799d2c5bb0>,\n",
       " <__main__.Protein at 0x2799d2c5c10>,\n",
       " <__main__.Protein at 0x2799d2c5c70>,\n",
       " <__main__.Protein at 0x2799d2c5cd0>,\n",
       " <__main__.Protein at 0x2799d2c5d30>,\n",
       " <__main__.Protein at 0x2799d2c5d90>,\n",
       " <__main__.Protein at 0x2799d2c5df0>,\n",
       " <__main__.Protein at 0x2799d2c5e50>,\n",
       " <__main__.Protein at 0x2799d2c5eb0>,\n",
       " <__main__.Protein at 0x2799d2c5f10>,\n",
       " <__main__.Protein at 0x2799d2c5f70>,\n",
       " <__main__.Protein at 0x2799d2c5fd0>,\n",
       " <__main__.Protein at 0x2799d2d7070>,\n",
       " <__main__.Protein at 0x2799d2d70d0>,\n",
       " <__main__.Protein at 0x2799d2d7130>,\n",
       " <__main__.Protein at 0x2799d2d7190>,\n",
       " <__main__.Protein at 0x2799d2d71f0>,\n",
       " <__main__.Protein at 0x2799d2d7250>,\n",
       " <__main__.Protein at 0x2799d2d72b0>,\n",
       " <__main__.Protein at 0x2799d2d7310>,\n",
       " <__main__.Protein at 0x2799d2d7370>,\n",
       " <__main__.Protein at 0x2799d2d73d0>,\n",
       " <__main__.Protein at 0x2799d2d7430>,\n",
       " <__main__.Protein at 0x2799d2d7490>,\n",
       " <__main__.Protein at 0x2799d2d74f0>,\n",
       " <__main__.Protein at 0x2799d2d7550>,\n",
       " <__main__.Protein at 0x2799d2d75b0>,\n",
       " <__main__.Protein at 0x2799d2d7610>,\n",
       " <__main__.Protein at 0x2799d2d7670>,\n",
       " <__main__.Protein at 0x2799d2d76d0>,\n",
       " <__main__.Protein at 0x2799d2d7730>,\n",
       " <__main__.Protein at 0x2799d2d7790>,\n",
       " <__main__.Protein at 0x2799d2d77f0>,\n",
       " <__main__.Protein at 0x2799d2d7850>,\n",
       " <__main__.Protein at 0x2799d2d78b0>,\n",
       " <__main__.Protein at 0x2799d2d7910>,\n",
       " <__main__.Protein at 0x2799d2d7970>,\n",
       " <__main__.Protein at 0x2799d2d79d0>,\n",
       " <__main__.Protein at 0x2799d2d7a30>,\n",
       " <__main__.Protein at 0x2799d2d7a90>,\n",
       " <__main__.Protein at 0x2799d2d7af0>,\n",
       " <__main__.Protein at 0x2799d2d7b50>,\n",
       " <__main__.Protein at 0x2799d2d7bb0>,\n",
       " <__main__.Protein at 0x2799d2d7c10>,\n",
       " <__main__.Protein at 0x2799d2d7c70>,\n",
       " <__main__.Protein at 0x2799d2d7cd0>,\n",
       " <__main__.Protein at 0x2799d2d7d30>,\n",
       " <__main__.Protein at 0x2799d2d7d90>,\n",
       " <__main__.Protein at 0x2799d2d7df0>,\n",
       " <__main__.Protein at 0x2799d2d7e50>,\n",
       " <__main__.Protein at 0x2799d2d7eb0>,\n",
       " <__main__.Protein at 0x2799d2d7f10>,\n",
       " <__main__.Protein at 0x2799d2d7f70>,\n",
       " <__main__.Protein at 0x2799d2d7fd0>,\n",
       " <__main__.Protein at 0x2799d2e7070>,\n",
       " <__main__.Protein at 0x2799d2e70d0>,\n",
       " <__main__.Protein at 0x2799d2e7130>,\n",
       " <__main__.Protein at 0x2799d2e7190>,\n",
       " <__main__.Protein at 0x2799d2e71f0>,\n",
       " <__main__.Protein at 0x2799d2e7250>,\n",
       " <__main__.Protein at 0x2799d2e72b0>,\n",
       " <__main__.Protein at 0x2799d2e7310>,\n",
       " <__main__.Protein at 0x2799d2e7370>,\n",
       " <__main__.Protein at 0x2799d2e73d0>,\n",
       " <__main__.Protein at 0x2799d2e7430>,\n",
       " <__main__.Protein at 0x2799d2e7490>,\n",
       " <__main__.Protein at 0x2799d2e74f0>,\n",
       " <__main__.Protein at 0x2799d2e7550>,\n",
       " <__main__.Protein at 0x2799d2e75b0>,\n",
       " <__main__.Protein at 0x2799d2e7610>,\n",
       " <__main__.Protein at 0x2799d2e7670>,\n",
       " <__main__.Protein at 0x2799d2e76d0>,\n",
       " <__main__.Protein at 0x2799d2e7730>,\n",
       " <__main__.Protein at 0x2799d2e7790>,\n",
       " <__main__.Protein at 0x2799d2e77f0>,\n",
       " <__main__.Protein at 0x2799d2e7850>,\n",
       " <__main__.Protein at 0x2799d2e78b0>,\n",
       " <__main__.Protein at 0x2799d2e7910>,\n",
       " <__main__.Protein at 0x2799d2e7970>,\n",
       " <__main__.Protein at 0x2799d2e79d0>,\n",
       " <__main__.Protein at 0x2799d2e7a30>,\n",
       " <__main__.Protein at 0x2799d2e7a90>,\n",
       " <__main__.Protein at 0x2799d2e7af0>,\n",
       " <__main__.Protein at 0x2799d2e7b50>,\n",
       " <__main__.Protein at 0x2799d2e7bb0>,\n",
       " <__main__.Protein at 0x2799d2e7c10>,\n",
       " <__main__.Protein at 0x2799d2e7c70>,\n",
       " ...]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x=Menu(menu,\"Gene\",\"Peptide\",\"UniprotID\")\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALDH2.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALDH2.identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MTTSQKHRDFVAEPMGEKPVGSLAGIGEVLGKKLEERGFDKAYVVLGQFLVLKKDEDLFREWLKDTCGANAKQSRDCFGCLREWCDAFL'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence=menu.loc[7421,\"Peptide\"]\n",
    "sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_rules(rules,new_key,new_value):\n",
    "    dictionary.update(new_key=new_value)\n",
    "    return rules\n",
    "    \n",
    "rules = {\n",
    "    'arg-c':         r'R',\n",
    "    'asp-n':         r'\\w(?=D)',\n",
    "    'bnps-skatole' : r'W',\n",
    "    'caspase 1':     r'(?<=[FWYL]\\w[HAT])D(?=[^PEDQKR])',\n",
    "    'caspase 2':     r'(?<=DVA)D(?=[^PEDQKR])',\n",
    "    'caspase 3':     r'(?<=DMQ)D(?=[^PEDQKR])',\n",
    "    'caspase 4':     r'(?<=LEV)D(?=[^PEDQKR])',\n",
    "    'caspase 5':     r'(?<=[LW]EH)D',\n",
    "    'caspase 6':     r'(?<=VE[HI])D(?=[^PEDQKR])',\n",
    "    'caspase 7':     r'(?<=DEV)D(?=[^PEDQKR])',\n",
    "    'caspase 8':     r'(?<=[IL]ET)D(?=[^PEDQKR])',\n",
    "    'caspase 9':     r'(?<=LEH)D',\n",
    "    'caspase 10':    r'(?<=IEA)D',\n",
    "    'chymotrypsin high specificity' : r'([FY](?=[^P]))|(W(?=[^MP]))',\n",
    "    'chymotrypsin low specificity':\n",
    "        r'([FLY](?=[^P]))|(W(?=[^MP]))|(M(?=[^PY]))|(H(?=[^DMPW]))',\n",
    "    'clostripain':   r'R',\n",
    "    'cnbr':          r'M',\n",
    "    'enterokinase':  r'(?<=[DE]{3})K',\n",
    "    'factor xa':     r'(?<=[AFGILTVM][DE]G)R',\n",
    "    'formic acid':   r'D',\n",
    "    'glutamyl endopeptidase': r'E',\n",
    "    'granzyme b':    r'(?<=IEP)D',\n",
    "    'hydroxylamine': r'N(?=G)',\n",
    "    'iodosobenzoic acid': r'W',\n",
    "    'lysc':          r'K',\n",
    "    'ntcb':          r'\\w(?=C)',\n",
    "    'pepsin ph1.3':  r'((?<=[^HKR][^P])[^R](?=[FL][^P]))|'\n",
    "                     r'((?<=[^HKR][^P])[FL](?=\\w[^P]))',\n",
    "    'pepsin ph2.0':  r'((?<=[^HKR][^P])[^R](?=[FLWY][^P]))|'\n",
    "                     r'((?<=[^HKR][^P])[FLWY](?=\\w[^P]))',\n",
    "    'proline endopeptidase': r'(?<=[HKR])P(?=[^P])',\n",
    "    'proteinase k':  r'[AEFILTVWY]',\n",
    "    'staphylococcal peptidase i': r'(?<=[^E])E',\n",
    "    'thermolysin':   r'[^DE](?=[AFILMV])',\n",
    "    'thrombin':      r'((?<=G)R(?=G))|'\n",
    "                     r'((?<=[AFGILTVM][AFGILTVWA]P)R(?=[^DE][^DE]))',\n",
    "    'trypsin':       r'([KR](?=[^P]))|((?<=W)K(?=P))|((?<=M)R(?=P))',\n",
    "    'trypsin_exception': r'((?<=[CD])K(?=D))|((?<=C)K(?=[HY]))|((?<=C)R(?=K))|((?<=R)R(?=[HR]))',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Cleaver(sequence, rule, missed_cleavages=0, min_length=None, max_length=None, exception=None):\n",
    "    peptides = []\n",
    "    if rule in rules:\n",
    "        rule = rules[rule]\n",
    "    exception = rules.get(exception, exception)\n",
    "    ml = missed_cleavages + 2\n",
    "    trange = range(ml) #returns range of 0 to ml-1\n",
    "    cleavage_sites = deque([0], maxlen=ml) # returns positions of cleavage sites between each other. \n",
    "    if min_length is None:\n",
    "        min_length = 1\n",
    "    cl = 1\n",
    "    if exception is not None: #locates postion of c-terminal by-product of cleavage. \n",
    "        exceptions = {x.end() for x in re.finditer(exception, sequence)}\n",
    "    for i in it.chain([x.end() for x in re.finditer(rule, sequence)],\n",
    "                   [None]):\n",
    "        if exception is not None and i in exceptions:\n",
    "            continue\n",
    "        cleavage_sites.append(i)\n",
    "#         print(cleavage_sites)\n",
    "        if cl < ml:\n",
    "            cl += 1\n",
    "        for j in trange[:cl - 1]:\n",
    "            seq = sequence[cleavage_sites[j]:cleavage_sites[-1]]\n",
    "            if (seq and len(seq) >= min_length) & (len(seq) < max_length):\n",
    "                peptides.append(seq)\n",
    "    return peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DFVAEPMGEKPVGSLAGIGEVLGK', 'AYVVLGQFLVLK', 'DTCGANAK', 'DCFGCLR', 'EWCDAFL']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missed_cleavages=0\n",
    "min_length=7\n",
    "max_length=100\n",
    "y=Cleaver(sequence,rule=\"([KR](?=[^P]))|((?<=W)K(?=P))|((?<=M)R(?=P))\",exception=None,missed_cleavages=0,min_length=7,max_length=100)\n",
    "y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defined Variables, Compiliers, Accessory Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Accessory Functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PEAKS_Importer(csv,drop_OG=True):\n",
    "    df=pd.read_csv(csv)\n",
    "    df[[\"Protein\",\"Y\"]]=df['Protein Accession'].str.split(\"|\",1,expand=True)\n",
    "    df[[\"Gene\",\"Species\"]]=df['Y'].str.split(\"_\",1,expand=True)\n",
    "    if drop_OG==True:\n",
    "        df.drop(columns=['Y', 'Protein Accession'], inplace=True)\n",
    "    else:\n",
    "        df.drop(columns=['Y'], inplace=True)\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Excel_Mapper(list_dfs, xls_path):\n",
    "    with ExcelWriter(xls_path) as writer:\n",
    "        for n, df in enumerate(list_dfs):\n",
    "            df.to_excel(writer,'sheet%s' % n)\n",
    "        writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def MeatWrapper(workbook,old,new,output):\n",
    "    meat = openpyxl.load_workbook(workbook)\n",
    "    wrap = meat[old]\n",
    "    wrap.title = new\n",
    "    meat.save(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Sampler(df,toothpicks,samples):\n",
    "    platter = {}\n",
    "    for toothpick, sample in zip(toothpicks, samples):\n",
    "        platter[sample] = df[toothpick]\n",
    "    return platter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Primary Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Mass(peptide):\n",
    "    mass = {  \"A\": 71.037114,\n",
    "    \"R\": 156.101111,\n",
    "    \"N\": 114.042927,\n",
    "    \"D\": 115.026943,\n",
    "    \"C\": 103.009185,\n",
    "    \"Q\": 129.042593,\n",
    "    \"E\": 128.058578,\n",
    "    \"G\": 57.021464,\n",
    "    \"H\": 137.058912,\n",
    "    \"I\": 113.084064,\n",
    "    \"L\": 113.084064,\n",
    "    \"K\": 128.094963,\n",
    "    \"M\": 131.040485,\n",
    "    \"F\": 147.068414,\n",
    "    \"P\": 97.052764,\n",
    "    \"S\": 87.032028,\n",
    "    \"T\": 101.047679,\n",
    "    \"W\": 186.079313,\n",
    "    \"Y\": 163.06332,\n",
    "    \"V\": 99.068414,\n",
    "    }\n",
    "    mass_list = [mass.get(aa,0.0)for aa in peptide]\n",
    "    pep_mass=sum(mass_list)\n",
    "    return pep_mass "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ButcherShop(df,target,identifier,rule, min_length=7,exception=None,max_length=100, pH=2.0, min_charge=2,missed=0):\n",
    "    pep_dict = {}\n",
    "    pep_dict_list = []\n",
    "    string_catcher=re.compile(r'^([A-Z]+)$')\n",
    "    print(f'You order is being processed and the butcher is preparing your {rule}-cut protein(s)!')\n",
    "    print(\"The butcher is working...\")\n",
    "    raw=df[[target, identifier]].set_index(identifier).to_dict()[target]\n",
    "    for gene,peptide in raw.items():\n",
    "        pep_dict[gene] = Cleaver(peptide,rule=rule,min_length=min_length,exception=exception,missed_cleavages=missed, max_length=max_length)\n",
    "    \n",
    "    for k, lst in pep_dict.items():    \n",
    "        d = {}\n",
    "        for i in range(len(lst)):\n",
    "            d.update({k: lst[i]})\n",
    "            d.update({'gene':k})\n",
    "            d.update({'aa_comp': dict(parser.amino_acid_composition(lst[i]))})\n",
    "            d.update({'peptide': re.findall(string_catcher,lst[i])})\n",
    "            d.update({'Length': len(lst[i])})\n",
    "            d.update({'z': int(round(electrochem.charge(lst[i], pH=pH)))})\n",
    "            d.update({'Mass': int(Scales.Mass(lst[i]))})\n",
    "            if d[\"z\"] > 0:\n",
    "                d.update({'m/z': d[\"Mass\"]/d['z']})\n",
    "            new_d = d.copy()\n",
    "            pep_dict_list.append(new_d)\n",
    "    print(\"Preparing your order...\")\n",
    "    pep_dict_list = [peptide for peptide in pep_dict_list if peptide['z'] >= int(min_charge)]\n",
    "    print(f'Order is up! You have acquired {len(pep_dict_list)} peptides that are between {min_length} and {max_length} amino acids!')\n",
    "    return pep_dict_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Deli(z,meat_package=False):\n",
    "    #z = list of dictionaries, keys must be equal thus will drop keys which are not cosistent between dictionaries\n",
    "    # use after ButcherShop\n",
    "    #returns dataframe\n",
    "    key_intersect = set(z[0].keys()).intersection(set(z[1].keys()))\n",
    "    zz = [{key:value for (key,value) in dicts.items() if key in key_intersect} for dicts in z]\n",
    "    ham = pd.DataFrame(zz)\n",
    "    if meat_package == True:\n",
    "        ham_counts=ham.groupby('gene').size().reset_index(name='counts')\n",
    "        ham=ham.merge(ham_counts,how='left', on=['gene'])\n",
    "    ham.drop(ham.columns[0],axis=1,inplace=True)\n",
    "    columns=ham.columns.tolist()\n",
    "    r = re.compile(\"^[pP]\")\n",
    "    P = list(filter(r.match, columns)) \n",
    "    peptide =str(P[0])\n",
    "    ham[peptide]=ham[peptide].apply(\",\".join) # convert list within df to string\n",
    "    return ham"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You order is being processed and the butcher is preparing your trypsin-cut protein(s)!\n",
      "The butcher is working...\n",
      "Preparing your order...\n",
      "Order is up! You have acquired 2682689 peptides that are between 7 and 100 amino acids!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>aa_comp</th>\n",
       "      <th>peptide</th>\n",
       "      <th>Length</th>\n",
       "      <th>z</th>\n",
       "      <th>Mass</th>\n",
       "      <th>m/z</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HV404</td>\n",
       "      <td>{'M': 1, 'K': 1, 'H': 1, 'L': 4, 'W': 1, 'F': ...</td>\n",
       "      <td>MKHLWFFLLLVAAPR</td>\n",
       "      <td>15</td>\n",
       "      <td>4</td>\n",
       "      <td>1823</td>\n",
       "      <td>455.750000</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HV404</td>\n",
       "      <td>{'H': 1, 'L': 4, 'W': 1, 'F': 2, 'V': 1, 'A': ...</td>\n",
       "      <td>HLWFFLLLVAAPR</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>1563</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HV404</td>\n",
       "      <td>{'M': 1, 'K': 2, 'H': 1, 'L': 9, 'W': 5, 'F': ...</td>\n",
       "      <td>MKHLWFFLLLVAAPRWVLSQVQLQESGPGLVKPSGTLSLTCAVSGG...</td>\n",
       "      <td>58</td>\n",
       "      <td>6</td>\n",
       "      <td>6380</td>\n",
       "      <td>1063.333333</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HV404</td>\n",
       "      <td>{'H': 1, 'L': 9, 'W': 5, 'F': 2, 'V': 6, 'A': ...</td>\n",
       "      <td>HLWFFLLLVAAPRWVLSQVQLQESGPGLVKPSGTLSLTCAVSGGSI...</td>\n",
       "      <td>56</td>\n",
       "      <td>5</td>\n",
       "      <td>6121</td>\n",
       "      <td>1224.200000</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HV404</td>\n",
       "      <td>{'W': 4, 'V': 5, 'L': 5, 'S': 10, 'Q': 3, 'E':...</td>\n",
       "      <td>WVLSQVQLQESGPGLVKPSGTLSLTCAVSGGSISSSNWWSWVR</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "      <td>4557</td>\n",
       "      <td>1519.000000</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2682684</th>\n",
       "      <td>PKHA5</td>\n",
       "      <td>{'N': 2, 'S': 4, 'V': 4, 'D': 1, 'E': 5, 'Q': ...</td>\n",
       "      <td>NSVDEQEETVISYESTPEVSRGNQTMAVK</td>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>3205</td>\n",
       "      <td>1068.333333</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2682685</th>\n",
       "      <td>PKHA5</td>\n",
       "      <td>{'G': 1, 'N': 1, 'Q': 1, 'T': 1, 'M': 1, 'A': ...</td>\n",
       "      <td>GNQTMAVK</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>830</td>\n",
       "      <td>415.000000</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2682686</th>\n",
       "      <td>PKHA5</td>\n",
       "      <td>{'N': 2, 'S': 12, 'V': 6, 'D': 1, 'E': 7, 'Q':...</td>\n",
       "      <td>NSVDEQEETVISYESTPEVSRGNQTMAVKSLSPSPESSASPVPSTQ...</td>\n",
       "      <td>58</td>\n",
       "      <td>4</td>\n",
       "      <td>6173</td>\n",
       "      <td>1543.250000</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2682687</th>\n",
       "      <td>PKHA5</td>\n",
       "      <td>{'G': 2, 'N': 1, 'Q': 3, 'T': 3, 'M': 2, 'A': ...</td>\n",
       "      <td>GNQTMAVKSLSPSPESSASPVPSTQPQLTEGSHFMCV</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>3798</td>\n",
       "      <td>1266.000000</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2682688</th>\n",
       "      <td>PKHA5</td>\n",
       "      <td>{'S': 8, 'L': 2, 'P': 5, 'E': 2, 'A': 1, 'V': ...</td>\n",
       "      <td>SLSPSPESSASPVPSTQPQLTEGSHFMCV</td>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>2968</td>\n",
       "      <td>1484.000000</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2682689 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene                                            aa_comp  \\\n",
       "0        HV404  {'M': 1, 'K': 1, 'H': 1, 'L': 4, 'W': 1, 'F': ...   \n",
       "1        HV404  {'H': 1, 'L': 4, 'W': 1, 'F': 2, 'V': 1, 'A': ...   \n",
       "2        HV404  {'M': 1, 'K': 2, 'H': 1, 'L': 9, 'W': 5, 'F': ...   \n",
       "3        HV404  {'H': 1, 'L': 9, 'W': 5, 'F': 2, 'V': 6, 'A': ...   \n",
       "4        HV404  {'W': 4, 'V': 5, 'L': 5, 'S': 10, 'Q': 3, 'E':...   \n",
       "...        ...                                                ...   \n",
       "2682684  PKHA5  {'N': 2, 'S': 4, 'V': 4, 'D': 1, 'E': 5, 'Q': ...   \n",
       "2682685  PKHA5  {'G': 1, 'N': 1, 'Q': 1, 'T': 1, 'M': 1, 'A': ...   \n",
       "2682686  PKHA5  {'N': 2, 'S': 12, 'V': 6, 'D': 1, 'E': 7, 'Q':...   \n",
       "2682687  PKHA5  {'G': 2, 'N': 1, 'Q': 3, 'T': 3, 'M': 2, 'A': ...   \n",
       "2682688  PKHA5  {'S': 8, 'L': 2, 'P': 5, 'E': 2, 'A': 1, 'V': ...   \n",
       "\n",
       "                                                   peptide  Length  z  Mass  \\\n",
       "0                                          MKHLWFFLLLVAAPR      15  4  1823   \n",
       "1                                            HLWFFLLLVAAPR      13  3  1563   \n",
       "2        MKHLWFFLLLVAAPRWVLSQVQLQESGPGLVKPSGTLSLTCAVSGG...      58  6  6380   \n",
       "3        HLWFFLLLVAAPRWVLSQVQLQESGPGLVKPSGTLSLTCAVSGGSI...      56  5  6121   \n",
       "4              WVLSQVQLQESGPGLVKPSGTLSLTCAVSGGSISSSNWWSWVR      43  3  4557   \n",
       "...                                                    ...     ... ..   ...   \n",
       "2682684                      NSVDEQEETVISYESTPEVSRGNQTMAVK      29  3  3205   \n",
       "2682685                                           GNQTMAVK       8  2   830   \n",
       "2682686  NSVDEQEETVISYESTPEVSRGNQTMAVKSLSPSPESSASPVPSTQ...      58  4  6173   \n",
       "2682687              GNQTMAVKSLSPSPESSASPVPSTQPQLTEGSHFMCV      37  3  3798   \n",
       "2682688                      SLSPSPESSASPVPSTQPQLTEGSHFMCV      29  2  2968   \n",
       "\n",
       "                 m/z  counts  \n",
       "0         455.750000      22  \n",
       "1         521.000000      22  \n",
       "2        1063.333333      22  \n",
       "3        1224.200000      22  \n",
       "4        1519.000000      22  \n",
       "...              ...     ...  \n",
       "2682684  1068.333333     312  \n",
       "2682685   415.000000     312  \n",
       "2682686  1543.250000     312  \n",
       "2682687  1266.000000     312  \n",
       "2682688  1484.000000     312  \n",
       "\n",
       "[2682689 rows x 8 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "z=ButcherShop(menu,target=\"Peptide\",identifier=\"Gene\", rule=\"trypsin\",missed=2,exception=None,min_length=7,max_length=100)\n",
    "z=Deli(z,meat_package=True)\n",
    "z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "meat=ButcherShop(menu,\"Peptide\",\"Gene\",\"thermolysin\")\n",
    "cuts=DeliShop(meat,meat_package=True)\n",
    "cuts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Origin(origin,target,Origin_Key=None,Target_Key=None,Origin_Label=None, Target_Label=None):\n",
    "    # primarily used to transfer whole protein measurments (i.e. aa_length) to data_frame with peptide products. Origin_Key is\n",
    "    # is used to find 'Merge' points (Target_Key) regardless of frequency (likely multiple peptides for one key/protein).Label is used\n",
    "    # capture information (peptide lenght, sequence, etc) from origin to target. New_Label is optional, default uses origin label to\n",
    "    # label newly created target column. \n",
    "    labeler=origin.set_index(Origin_Key).to_dict()[Origin_Label]\n",
    "    if Target_Label == None:\n",
    "        target[Origin_Label] = target[Target_Key].map(labeler)\n",
    "    else:\n",
    "        target[Target_Label] = target[Target_Key].map(labeler)\n",
    "    return target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>aa_comp</th>\n",
       "      <th>peptide</th>\n",
       "      <th>Length</th>\n",
       "      <th>z</th>\n",
       "      <th>Mass</th>\n",
       "      <th>m/z</th>\n",
       "      <th>counts</th>\n",
       "      <th>OG_LENGTH</th>\n",
       "      <th>% of Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1047460</th>\n",
       "      <td>DWORF</td>\n",
       "      <td>{'M': 2, 'A': 2, 'E': 1, 'K': 1, 'G': 3, 'S': ...</td>\n",
       "      <td>MAEKAGSTFSHLLVPILLLIGWIVGCIIMIYVVFS</td>\n",
       "      <td>35</td>\n",
       "      <td>3</td>\n",
       "      <td>3814</td>\n",
       "      <td>1271.333333</td>\n",
       "      <td>2</td>\n",
       "      <td>35</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>TJB15</td>\n",
       "      <td>{'S': 2, 'N': 1, 'Q': 2, 'P': 1, 'H': 1, 'F': ...</td>\n",
       "      <td>SNQPQHFGDGTRLSIL</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>1752</td>\n",
       "      <td>584.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1134532</th>\n",
       "      <td>TJA42</td>\n",
       "      <td>{'Y': 1, 'G': 5, 'S': 2, 'Q': 1, 'N': 1, 'L': ...</td>\n",
       "      <td>YGGSQGNLIFGKGTKLSVKP</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>2033</td>\n",
       "      <td>508.250000</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171870</th>\n",
       "      <td>OST4</td>\n",
       "      <td>{'M': 2, 'I': 2, 'T': 1, 'D': 1, 'V': 6, 'Q': ...</td>\n",
       "      <td>MITDVQLAIFANMLGVSLFLLVVLYHYVAVNNPKKQE</td>\n",
       "      <td>37</td>\n",
       "      <td>4</td>\n",
       "      <td>4173</td>\n",
       "      <td>1043.250000</td>\n",
       "      <td>3</td>\n",
       "      <td>37</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398799</th>\n",
       "      <td>BAGE4</td>\n",
       "      <td>{'M': 2, 'A': 7, 'G': 2, 'V': 3, 'F': 1, 'L': ...</td>\n",
       "      <td>MAAGAVFLALSAQLLQARLMKEESPVVSWWLEPEDGTAL</td>\n",
       "      <td>39</td>\n",
       "      <td>3</td>\n",
       "      <td>4207</td>\n",
       "      <td>1402.333333</td>\n",
       "      <td>4</td>\n",
       "      <td>39</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324874</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'R': 1, 'E': 1, 'P': 1, 'V': 1, 'A': 1, 'I': ...</td>\n",
       "      <td>REPVAIK</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>792</td>\n",
       "      <td>264.000000</td>\n",
       "      <td>9507</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324423</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'Y': 1, 'R': 2, 'M': 1, 'H': 1, 'S': 1, 'D': 1}</td>\n",
       "      <td>YRMHSDR</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>945</td>\n",
       "      <td>236.250000</td>\n",
       "      <td>9507</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332177</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'S': 1, 'A': 1, 'T': 1, 'V': 2, 'L': 1, 'K': 1}</td>\n",
       "      <td>SATVLVK</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>698</td>\n",
       "      <td>349.000000</td>\n",
       "      <td>9507</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326407</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'E': 1, 'V': 2, 'I': 1, 'Q': 2, 'K': 1}</td>\n",
       "      <td>EVIQVQK</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>825</td>\n",
       "      <td>412.500000</td>\n",
       "      <td>9507</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326991</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'E': 1, 'K': 2, 'I': 1, 'F': 1, 'Q': 1, 'L': 1}</td>\n",
       "      <td>EKIFQLK</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>886</td>\n",
       "      <td>295.333333</td>\n",
       "      <td>9507</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2682689 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene                                            aa_comp  \\\n",
       "1047460  DWORF  {'M': 2, 'A': 2, 'E': 1, 'K': 1, 'G': 3, 'S': ...   \n",
       "128      TJB15  {'S': 2, 'N': 1, 'Q': 2, 'P': 1, 'H': 1, 'F': ...   \n",
       "1134532  TJA42  {'Y': 1, 'G': 5, 'S': 2, 'Q': 1, 'N': 1, 'L': ...   \n",
       "171870    OST4  {'M': 2, 'I': 2, 'T': 1, 'D': 1, 'V': 6, 'Q': ...   \n",
       "398799   BAGE4  {'M': 2, 'A': 7, 'G': 2, 'V': 3, 'F': 1, 'L': ...   \n",
       "...        ...                                                ...   \n",
       "324874   TITIN  {'R': 1, 'E': 1, 'P': 1, 'V': 1, 'A': 1, 'I': ...   \n",
       "324423   TITIN   {'Y': 1, 'R': 2, 'M': 1, 'H': 1, 'S': 1, 'D': 1}   \n",
       "332177   TITIN   {'S': 1, 'A': 1, 'T': 1, 'V': 2, 'L': 1, 'K': 1}   \n",
       "326407   TITIN           {'E': 1, 'V': 2, 'I': 1, 'Q': 2, 'K': 1}   \n",
       "326991   TITIN   {'E': 1, 'K': 2, 'I': 1, 'F': 1, 'Q': 1, 'L': 1}   \n",
       "\n",
       "                                         peptide  Length  z  Mass  \\\n",
       "1047460      MAEKAGSTFSHLLVPILLLIGWIVGCIIMIYVVFS      35  3  3814   \n",
       "128                             SNQPQHFGDGTRLSIL      16  3  1752   \n",
       "1134532                     YGGSQGNLIFGKGTKLSVKP      20  4  2033   \n",
       "171870     MITDVQLAIFANMLGVSLFLLVVLYHYVAVNNPKKQE      37  4  4173   \n",
       "398799   MAAGAVFLALSAQLLQARLMKEESPVVSWWLEPEDGTAL      39  3  4207   \n",
       "...                                          ...     ... ..   ...   \n",
       "324874                                   REPVAIK       7  3   792   \n",
       "324423                                   YRMHSDR       7  4   945   \n",
       "332177                                   SATVLVK       7  2   698   \n",
       "326407                                   EVIQVQK       7  2   825   \n",
       "326991                                   EKIFQLK       7  3   886   \n",
       "\n",
       "                 m/z  counts  OG_LENGTH  % of Protein  \n",
       "1047460  1271.333333       2         35    100.000000  \n",
       "128       584.000000       2         16    100.000000  \n",
       "1134532   508.250000       4         20    100.000000  \n",
       "171870   1043.250000       3         37    100.000000  \n",
       "398799   1402.333333       4         39    100.000000  \n",
       "...              ...     ...        ...           ...  \n",
       "324874    264.000000    9507      34350      0.020378  \n",
       "324423    236.250000    9507      34350      0.020378  \n",
       "332177    349.000000    9507      34350      0.020378  \n",
       "326407    412.500000    9507      34350      0.020378  \n",
       "326991    295.333333    9507      34350      0.020378  \n",
       "\n",
       "[2682689 rows x 10 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fresh=Peptide_Origin(menu,z,Origin_Key=\"Gene\",Target_Key=\"gene\",Origin_Label=\"Protein_Length\",Target_Label=\"OG_LENGTH\")\n",
    "fresh[\"% of Protein\"]= (fresh[\"Length\"]/fresh[\"OG_LENGTH\"])*100\n",
    "fresh.sort_values(by=[\"% of Protein\"],ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICE_cloud(df,x=0,target=\"aa_comp\"):\n",
    "    word_could_dict = df[target][x]\n",
    "    wordcloud = WordCloud(width = 500, height = 500,scale=10,prefer_horizontal=1,relative_scaling=1,min_font_size=18,max_font_size=48,font_step=8,\n",
    "                      background_color='white',contour_width=1,contour_color=\"black\" ).generate_from_frequencies(word_could_dict)\n",
    "    plt.figure(figsize=(15,8))\n",
    "    plt.imshow(wordcloud)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Handles Up to 3 Replicates (t_id1-3) per Selection, used for removing amino acids from N-terminal during PEAKS exports with enzyme that cleave at C-terminal. \n",
    "#Can be used to remove M from N-terminal of peptides produced by enzymes which cleave at N-terminal of target aa. \n",
    "def Butcher(df,ident1=None,ident2=None,ident3=None,t_id1=None,t_id2=None,t_id3=None,t_value=0,acid=[\"J\",\"Z\"],labels=list(),excel_mapper=True,excel_name=None):\n",
    "    raw=df.loc[:,df.columns.str.contains(ident1)]\n",
    "    tag=df[labels]\n",
    "    raw = pd.concat([raw, tag], axis=1)\n",
    "    raw[\"Peptide\"]= raw[\"Peptide\"].str.replace('\\W+',\"\")\n",
    "    raw[\"Peptide\"]= raw[\"Peptide\"].str.replace('\\d+',\"\")\n",
    "    raw[\"Peptide\"]= raw[\"Peptide\"].apply(lambda x : x[1:] if x.startswith(tuple(acid)) else x)\n",
    "    cuts=raw.loc[:,raw.columns.str.contains(ident2)]\n",
    "    if ident3 != None:\n",
    "        cuts=cuts.loc[:,cuts.columns.str.contains(ident3)]\n",
    "    cuts = pd.concat([cut, tag], axis=1)\n",
    "    blade = cuts.filter(regex=r'^AREA').isin(['0']).all(axis=1)\n",
    "    cuts=cuts.loc[~blade]\n",
    "    cuts.reset_index(inplace=True)\n",
    "    excels=[raw,cuts]\n",
    "    Excel_Mapper(excels,excel_name +\".xlsx\")\n",
    "    return raw, cuts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICE_bar(df,x=0,target=None,lock=None,key=None):\n",
    "    test_dict=df.loc[df[lock] == key][target].iloc[x]\n",
    "    x=list(test_dict.keys())\n",
    "    y=list(test_dict.values())\n",
    "    plt.bar(x,y, color='g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICE_plot(df,x=0,target=None,lock=None,key=None):\n",
    "    test_dict=df.loc[df[lock] == key][target].iloc[x]\n",
    "    x=list(test_dict.keys())\n",
    "    y=list(test_dict.values())\n",
    "    plt.plot(x,y, color='g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "single positional indexer is out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-61-4ed3b24bd64f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mPeptide_ICE_plot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfresh\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtarget\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"aa_comp\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlock\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"gene\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"LHDA\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-60-8374713b45df>\u001b[0m in \u001b[0;36mPeptide_ICE_plot\u001b[1;34m(df, x, target, lock, key)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mPeptide_ICE_plot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtarget\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mlock\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m     \u001b[0mtest_dict\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mlock\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mtarget\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m     \u001b[0mx\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtest_dict\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m     \u001b[0my\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtest_dict\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mplt\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcolor\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'g'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\SciFly\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   1766\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1767\u001b[0m             \u001b[0mmaybe_callable\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1768\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmaybe_callable\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1769\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1770\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\SciFly\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   2136\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2137\u001b[0m             \u001b[1;31m# validate the location\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2138\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_validate_integer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2139\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2140\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\SciFly\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_validate_integer\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   2061\u001b[0m         \u001b[0mlen_axis\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2062\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mkey\u001b[0m \u001b[1;33m>=\u001b[0m \u001b[0mlen_axis\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mkey\u001b[0m \u001b[1;33m<\u001b[0m \u001b[1;33m-\u001b[0m\u001b[0mlen_axis\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2063\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"single positional indexer is out-of-bounds\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2064\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2065\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_getitem_tuple\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: single positional indexer is out-of-bounds"
     ]
    }
   ],
   "source": [
    "Peptide_ICE_plot(fresh,x=0,target=\"aa_comp\", lock=\"gene\",key=\"LHDA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Wash(peptide,ccut=None, PEAKS=False):\n",
    "    if ccut is None:\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "    elif ccut == \"Trypsin\":\n",
    "        x=[ 'R', 'K']\n",
    "        if PEAKS:\n",
    "            peptide= peptide.apply(lambda x : x[1:] if x.startswith(tuple(x)) else x)\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "        return(peptide)\n",
    "    elif ccut == \"AspN\":\n",
    "        x=['A', 'R', 'N','D', 'C', 'Q','E', 'G', 'H',\n",
    "             'I', 'L', 'K', 'F', 'P',\n",
    "             'S', 'T', 'W','Y', 'V']\n",
    "        if PEAKS:\n",
    "            peptide= peptide.apply(lambda x : x[1:] if x.startswith(tuple(x)) else x)\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "        return(peptide)\n",
    "    elif ccut == \"GluC\":\n",
    "        x=['E']\n",
    "        if PEAKS:\n",
    "            peptide= peptide.apply(lambda x : x[1:] if x.startswith(tuple(x)) else x)\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "        return(peptide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_IPC(peptide,start_pH=6.51,Epsilon=0.01,):\n",
    "    IPC_score={'Cterm': 2.383, 'pKAsp': 3.887, 'pKGlu': 4.317, 'pKCys': 8.297, 'pKTyr': 10.071, 'pk_his': 6.018, 'Nterm': 9.564, 'pKLys': 10.517, 'pKArg': 12.503}\n",
    "    pKCterm = IPC_score['Cterm']\n",
    "    pKAsp = IPC_score['pKAsp']\n",
    "    pKGlu = IPC_score['pKGlu']\n",
    "    pKCys = IPC_score['pKCys']\n",
    "    pKTyr = IPC_score['pKTyr']\n",
    "    pKHis = IPC_score['pk_his']\n",
    "    pKNterm = IPC_score['Nterm']\n",
    "    pKLys = IPC_score['pKLys'] \n",
    "    pKArg = IPC_score['pKArg']\n",
    "    pH = start_pH      \n",
    "    pHprev = 0.0         \n",
    "    pHnext = 14.0        \n",
    "    E = Epsilon  \n",
    "    temp = 0.01\n",
    "    nterm=peptide[0]\n",
    "    cterm=peptide[-1]\n",
    "#will now cycle through all peptides until a pH within the epsilon is found       \n",
    "    while 1:             \n",
    "        QN1=-1.0/(1.0+pow(10,(pKCterm-pH)))                                        \n",
    "        QN2=-peptide.count('D')/(1.0+pow(10,(pKAsp-pH)))           \n",
    "        QN3=-peptide.count('E')/(1.0+pow(10,(pKGlu-pH)))           \n",
    "        QN4=-peptide.count('C')/(1.0+pow(10,(pKCys-pH)))           \n",
    "        QN5=-peptide.count('Y')/(1.0+pow(10,(pKTyr-pH)))        \n",
    "        QP1=peptide.count('H')/(1.0+pow(10,(pH-pKHis)))            \n",
    "        QP2=1.0/(1.0+pow(10,(pH-pKNterm)))                \n",
    "        QP3=peptide.count('K')/(1.0+pow(10,(pH-pKLys)))           \n",
    "        QP4=peptide.count('R')/(1.0+pow(10,(pH-pKArg)))            \n",
    "        NQ=QN1+QN2+QN3+QN4+QN5+QP1+QP2+QP3+QP4\n",
    "        \n",
    "        if NQ<0.0:                                  \n",
    "            temp = pH\n",
    "            pH = pH-((pH-pHprev)/2.0)\n",
    "            pHnext = temp\n",
    "\n",
    "        else:\n",
    "            temp = pH\n",
    "            pH = pH + ((pHnext-pH)/2.0)\n",
    "            pHprev = temp\n",
    "#terminal condition, finding pI with given precision defined by Epsilon\n",
    "        if (pH-pHprev<E) and (pHnext-pH<E): \n",
    "            return pH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICEmaker(peptide,excel_name=\"Test\",freeze_it=0,cut_it=6):\n",
    "    ice=peptide.str.slice(freeze_it,cut_it)\n",
    "    ice_ice = ice.str.slice(freeze_it,cut_it)\n",
    "    ice_ice.to_excel('ICE '+excel_name+\".xlsx\")\n",
    "    return ice_ice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Neutral_pH(peptide):\n",
    "    z_dict = {'E': -1, 'D': -1, 'K': 1, 'R': 1} \n",
    "    charge = [z_dict.get(aa, 0.0) for aa in peptide]\n",
    "    spark=sum(charge)\n",
    "    return(spark)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Gravy(peptide):\n",
    "    hydro = {     \"A\": 1.800,\n",
    "    \"R\": -4.500,\n",
    "    \"N\": -3.500,\n",
    "    \"D\": -3.500,\n",
    "    \"C\": 2.500,\n",
    "    \"Q\": -3.500,\n",
    "    \"E\": -3.500,\n",
    "    \"G\": -0.400,\n",
    "    \"H\": -3.200,\n",
    "    \"I\": 4.500,\n",
    "    \"L\": 3.800,\n",
    "    \"K\": -3.900,\n",
    "    \"M\": 1.900,\n",
    "    \"F\": 2.800,\n",
    "    \"P\": -1.600,\n",
    "    \"S\": -0.800,\n",
    "    \"T\": -0.700,\n",
    "    \"W\": -0.900,\n",
    "    \"Y\": -1.300,\n",
    "    \"V\": 4.200,\n",
    "    }\n",
    "    hydro_list = [hydro.get(aa,0.0)for aa in peptide]\n",
    "    hydro_sum=sum(hydro_list)\n",
    "    gravy=hydro_sum/len(hydro_list)\n",
    "    return gravy "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Inspector(df,target,new_column=None):\n",
    "    all_data = []\n",
    "    for peptide in df[target]:\n",
    "        temp=dict(parser.amino_acid_composition(peptide))\n",
    "        all_data.append(temp)\n",
    "    df[new_column]=all_data\n",
    "    print(f\"Peptide inspection is completed! Dictionaries are stored under {new_column} in dataframe!\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Marinate (df,target,length,IPC=False,Hydro=False,GRAVY=False,NeutralZ=False,Peptide_Inspector=False):\n",
    "    print(\"Marinating peptides...\")\n",
    "    if IPC is True:\n",
    "        df[\"IPC\"]=df[target].apply(Peptide_IPC)\n",
    "        print(\"IPC calculated!\")\n",
    "    if GRAVY is True:\n",
    "        df[\"Hydro_Sum\"]=df[target].apply(Peptide_Gravy)\n",
    "        print(\"Its all GRAVY Baby!\")\n",
    "    if NeutralZ is True:\n",
    "        df[\"Neutral_Z\"]=df[target].apply(Peptide_Neutral_pH)\n",
    "        print(\"Charge at Neutral pH added!\")\n",
    "    print(\"Peptides have been marinated!\")\n",
    "    return df\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Wishbone(df,bone,split):\n",
    "    x=df[bone].value_counts()\n",
    "    x2=pd.DataFrame(x,columns=[bone])\n",
    "    x3= x2.loc[(x2[bone])>int(split)]\n",
    "    x3z=len(x3)\n",
    "    x4=x3z/len(x2)*100\n",
    "    print(f\"The number of proteins with > {split} peptides : %.1f\" % x3z)\n",
    "    print(f\"Ratio of Proteins with > {split} Peptides Identified: %.3f\" % x4)\n",
    "    return x3,x3z,x4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Sweet_N_Sour(set1,set2):\n",
    "    common_IDs=list(set(set1)&set(set2))\n",
    "    common=len(common_IDs)\n",
    "    unique=len(list(set(set1)^set(set2)))\n",
    "    unique_set1=list(set(set1)-set(set2))\n",
    "    unique_set2=list(set(set2)-set(set1))\n",
    "    set1_count=len(unique_set1)\n",
    "    set2_count=len(unique_set2)\n",
    "    Ratio=(unique/(common+unique))*100\n",
    "    print(f\"The number of common peptides is: %.3f\" % common)\n",
    "    print(f\"The number of unique peptides in set1 is: %.3f\" % set1_count)\n",
    "    print(f\"The number of unique peptides in set1 is: %.3f\" % set2_count)\n",
    "    print(f\"The Ratio of Unique to Common proteins is: %.3f\" % Ratio)\n",
    "    return common_IDs,unique_set1, unique_set2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CorrPie(df,apples,figsize=(16,8),cmap='Blues',title=\"Title\",fontsize=18,pad=16,save_name=\"Heat_Test\",dpi=600,fmt='eps'):\n",
    "    og_corr=df[apples]\n",
    "    plt.figure(figsize=figsize)\n",
    "    mask=np.triu(np.ones_like(og_corr.corr(), dtype=np.bool))\n",
    "    heatmap=heatmap = sns.heatmap(og_corr.corr(), mask=mask, vmin=-1, vmax=1, annot=True, cmap=cmap)\n",
    "    heatmap.set_title(title, fontdict={'fontsize':fontsize}, pad=pad)\n",
    "    plt.savefig(save_name,format=fmt,dpi=dpi,bbox_inches=\"tight\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def WeddingCake(df,x,y,z,dpi=300,s=12,alpha=0.3,edgecolor='k',color=\"blue\",my_viewx=20,my_viewy=50,\n",
    "               xlabel=\"Parameter xlabel\",ylabel=\"Parameter ylabel\",zlabel=\"Parameter zlabel\",fmt='png',\n",
    "               figx=10,figy=10,xmin=0,xmax=1000,ymin=0,ymax=1000,zmin=0,zmax=1000):\n",
    "    fig = plt.figure(figsize=(figx,figy),dpi=dpi)\n",
    "    ax = fig.gca(projection='3d')\n",
    "    X = df[x]\n",
    "    Y = df[y]\n",
    "    Z = df[z]\n",
    "    ax.scatter(X,Y,Z,color=color,s=s,alpha=alpha,edgecolor=edgecolor)\n",
    "    ax.view_init(my_viewx,my_viewy)\n",
    "    ax.set_xlabel(xlabel)\n",
    "    ax.set_ylabel(ylabel)\n",
    "    ax.set_zlabel(zlabel)\n",
    "    ax.xaxis._axinfo[\"grid\"].update({\"linewidth\":1, \"color\" : \"grey\"})\n",
    "    ax.yaxis._axinfo[\"grid\"].update({\"linewidth\":1, \"color\" : \"grey\"})\n",
    "    ax.zaxis._axinfo[\"grid\"]['color'] = \"k\"\n",
    "    ax.zaxis._axinfo[\"grid\"]['linestyle'] = \"--\"\n",
    "    ax.set_ylim3d(ymin,ymax)\n",
    "    ax.set_xlim3d(xmin,xmax)\n",
    "    ax.set_zlim3d(zmin,zmax)\n",
    "    FigTitle=input('Figure Title:')\n",
    "    plt.savefig(FigTitle,format=fmt,dpi=dpi,bbox_inches=\"tight\")\n",
    "    plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Cake(df,x,y,dpi=600,s=25,alpha=0.25,edgecolor='k',color=\"blue\",fmt='eps',labsize=18,\n",
    "               figx=10,figy=10,xmin=0,xmax=1000,ymin=0,ymax=1000,xlabel=\"xlabel parameter\",ylabel=\"ylabel parameter\",\n",
    "        loc=2,pad=1,borderpad=1,frameon=True, show=False):\n",
    "    fig, ax = plt.subplots(1,1,figsize=(figx,figy),dpi=dpi)\n",
    "    X = df[x]\n",
    "    Y = df[y]\n",
    "    plt.scatter(X,Y,color=color,s=s, alpha=alpha,edgecolor=edgecolor)\n",
    "    plt.xlabel(xlabel, fontsize=labsize)\n",
    "    plt.ylabel(ylabel, fontsize=labsize)\n",
    "    corr, _ = pearsonr(X, Y)\n",
    "    # loc works the same as it does with figures (though best doesn't work)\n",
    "    # pad=5 will increase the size of padding between the border and text\n",
    "    # borderpad=5 will increase the distance between the border and the axes\n",
    "    # frameon=False will remove the box around the text\n",
    "    anchored_text = AnchoredText('Pearsons correlation: r = %.3f' % corr, loc=loc,pad=pad,borderpad=borderpad,frameon=frameon)\n",
    "    ax.add_artist(anchored_text)\n",
    "    FigTitle=input('Figure Title:')\n",
    "    plt.savefig(FigTitle,format=fmt,dpi=dpi,bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CakePop(df,x,xi,y,yi, figx=10,figy=10,dpi=600,font_scale=1.5,figstyle=\"white\",xlabel=\"xlabel\",ylabel=\"ylabel\",labsize=18,\n",
    "           cmapx=-0.3,cmapy=0.0,sizex=10,sizey=200,loc='upper left',pad=1,borderpad=1,frameon=True):\n",
    "    sns.set(font_scale=font_scale)\n",
    "    cake=sns.set_style(figstyle)\n",
    "    cmap = sns.cubehelix_palette(start=cmapx, rot=cmapy, as_cmap=True)\n",
    "    X = df[x]\n",
    "    Y = df[y]\n",
    "    cake=sns.relplot(data=df,\n",
    "    x=X, y=Y,\n",
    "    hue=xi, size=yi,\n",
    "    palette=cmap, sizes=(sizex, sizey))\n",
    "    ax = cake.axes[0,0]\n",
    "    ####\n",
    "    corr, _ = pearsonr(X, Y)\n",
    "    anchored_text = AnchoredText('Pearsons correlation: r = %.3f' % corr, loc=loc, prop=dict(size=labsize*0.5),pad=pad,borderpad=borderpad,frameon=frameon)\n",
    "    ax.add_artist(anchored_text)\n",
    "    ####\n",
    "    plt.xlabel(xlabel, fontsize=labsize)\n",
    "    plt.ylabel(ylabel, fontsize=labsize)\n",
    "    FigTitle=input('Figure Title:')\n",
    "    plt.savefig(FigTitle,format=fmt,dpi=dpi,bbox_inches=\"tight\")\n",
    "    plt.show(cake)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
