{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Peptide Chef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Created by: Tyler T. Cooper, PhD (2021)\n",
    "# Tools to Analyze Peptides and Amino Acid Compositions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Data Handling\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import ExcelWriter\n",
    "import openpyxl\n",
    "import re\n",
    "import pyteomics\n",
    "from pyteomics import fasta, parser, electrochem, mass\n",
    "from itertools import combinations\n",
    "from urllib.request import urlretrieve\n",
    "import gzip\n",
    "import itertools as it\n",
    "from collections import deque\n",
    "import functools\n",
    "from functools import lru_cache\n",
    "\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "\n",
    "#Figure Generation\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib import offsetbox\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "from matplotlib.ticker import NullFormatter\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from matplotlib.tri import Triangulation\n",
    "from mpl_toolkits.mplot3d import axes3d\n",
    "from IPython.display import Image, display\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "import glob\n",
    "import bioinfokit\n",
    "from bioinfokit import analys, visuz\n",
    "from wordcloud import WordCloud\n",
    "\n",
    "#Venn Diagrams\n",
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "from matplotlib_venn import venn3, venn3_circles\n",
    "%matplotlib inline\n",
    "\n",
    "cmap = 'PRGn'\n",
    "fmt='eps'\n",
    "dpi=600"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Fasta File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "url=\"https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/UP000005640_9606.fasta.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache(maxsize=None)\n",
    "def CookBook(Species=None,homebrew=False, takeout=True,url=None,measure=False,target=None):\n",
    "    #Use to import fasta.gz files directly from uniprot (Takeout) with a provided url (Default is Human) or from local folder using homebrew hyperparameter.\n",
    "    # Takeout is default, Takeout must be switched to False for Homebrew to be True.\n",
    "    # accepts Uniprot format\n",
    "    ingredients=list()\n",
    "    if url == None:\n",
    "        url=\"https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/UP000005640_9606.fasta.gz\"\n",
    "    if (homebrew == True) & (takeout==True):\n",
    "        print(\"Only one method (homebrew/takeout) can == True...\")\n",
    "        print (\"Takeout will be automatically selected...\")\n",
    "    if takeout is True:\n",
    "        homebrew=False\n",
    "        print('Downloading the FASTA file from url...')\n",
    "        urlretrieve(url,'temp.fasta.gz')\n",
    "        print('Unzipping...')\n",
    "        with gzip.open('temp.fasta.gz', mode='rt') as gzfile:\n",
    "            for info, contents in fasta.FASTA(gzfile):\n",
    "                taste=list((info,contents))\n",
    "                ingredients.append(taste)\n",
    "                recipie=pd.DataFrame(ingredients,columns=['ID','Peptide'])\n",
    "        print(\"Takeout is Done!\")\n",
    "    if homebrew is True:\n",
    "        if Species == None:\n",
    "            print(\"Provide Species identification. This will be the name of the local FASTA file.\")\n",
    "        print(\"Downloading the FASTA file from local flle...\")\n",
    "        book = str(Species+\".fasta\")\n",
    "        recipie=pd.DataFrame()\n",
    "        print(\"Serving up a homebrew...\")\n",
    "        with fasta.read(book) as menu:\n",
    "            for info, contents in menu:\n",
    "                taste=list((info,contents))\n",
    "                ingredients.append(taste)\n",
    "                recipie=pd.DataFrame(ingredients,columns=['ID','Peptide'])\n",
    "        print(\"Homebrow is Done!\")\n",
    "    recipie[['db', 'UniprotID','ID2']] = recipie['ID'].str.split('|', 2, expand=True)\n",
    "    recipie[['Gene','Identification']] = recipie['ID2'].str.split('_', 1, expand=True)\n",
    "    recipie.drop(columns=['ID', 'ID2',\"db\"], inplace=True)\n",
    "    if measure == True:\n",
    "        recipie[\"Protein_Length\"]=recipie[target].str.len()\n",
    "    print(\"Here ya go boss!\")\n",
    "    return(recipie)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading the FASTA file from url...\n",
      "Unzipping...\n",
      "Takeout is Done!\n",
      "Here ya go boss!\n"
     ]
    }
   ],
   "source": [
    "menu=CookBook(homebrew=False,takeout=True,url=url, measure=True, target=\"Peptide\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20614"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(menu.Gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Identification</th>\n",
       "      <th>Protein_Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6710</th>\n",
       "      <td>MTTSQKHRDFVAEPMGEKPVGSLAGIGEVLGKKLEERGFDKAYVVL...</td>\n",
       "      <td>O75531</td>\n",
       "      <td>BAF</td>\n",
       "      <td>HUMAN Barrier-to-autointegration factor OS=Hom...</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14012</th>\n",
       "      <td>MLRAAARFGPRLGRRLLSAAATQAVPAPNQQPEVFCNQIFINNEWH...</td>\n",
       "      <td>P05091</td>\n",
       "      <td>ALDH2</td>\n",
       "      <td>HUMAN Aldehyde dehydrogenase, mitochondrial OS...</td>\n",
       "      <td>517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20167</th>\n",
       "      <td>MAALGTVLFTGVRRLHCSVAAWAGGQWRLQQGLAANPSGYGPLTEL...</td>\n",
       "      <td>Q86TS9</td>\n",
       "      <td>RM52</td>\n",
       "      <td>HUMAN 39S ribosomal protein L52, mitochondrial...</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Peptide UniprotID   Gene  \\\n",
       "6710   MTTSQKHRDFVAEPMGEKPVGSLAGIGEVLGKKLEERGFDKAYVVL...    O75531    BAF   \n",
       "14012  MLRAAARFGPRLGRRLLSAAATQAVPAPNQQPEVFCNQIFINNEWH...    P05091  ALDH2   \n",
       "20167  MAALGTVLFTGVRRLHCSVAAWAGGQWRLQQGLAANPSGYGPLTEL...    Q86TS9   RM52   \n",
       "\n",
       "                                          Identification  Protein_Length  \n",
       "6710   HUMAN Barrier-to-autointegration factor OS=Hom...              89  \n",
       "14012  HUMAN Aldehyde dehydrogenase, mitochondrial OS...             517  \n",
       "20167  HUMAN 39S ribosomal protein L52, mitochondrial...             123  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lists=[\"ALDH2\",\"BAF\",\"RM52\"]\n",
    "menu.loc[menu['Gene'].isin(lists)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Scales:\n",
    "    def Mass(peptide):\n",
    "        mass = {  \n",
    "        \"A\": 71.037114,\n",
    "        \"R\": 156.101111,\n",
    "        \"N\": 114.042927,\n",
    "        \"D\": 115.026943,\n",
    "        \"C\": 103.009185,\n",
    "        \"Q\": 129.042593,\n",
    "        \"E\": 128.058578,\n",
    "        \"G\": 57.021464,\n",
    "        \"H\": 137.058912,\n",
    "        \"I\": 113.084064,\n",
    "        \"L\": 113.084064,\n",
    "        \"K\": 128.094963,\n",
    "        \"M\": 131.040485,\n",
    "        \"F\": 147.068414,\n",
    "        \"P\": 97.052764,\n",
    "        \"S\": 87.032028,\n",
    "        \"T\": 101.047679,\n",
    "        \"W\": 186.079313,\n",
    "        \"Y\": 163.06332,\n",
    "        \"V\": 99.068414,\n",
    "        }\n",
    "        mass_list = [mass.get(aa,0.0)for aa in peptide]\n",
    "        pep_mass=sum(mass_list)\n",
    "        return pep_mass\n",
    "    \n",
    "    def Peptide_IPC(peptide,start_pH=6.51,Epsilon=0.01,):\n",
    "        IPC_score={'Cterm': 2.383, 'pKAsp': 3.887, 'pKGlu': 4.317, 'pKCys': 8.297, 'pKTyr': 10.071, 'pk_his': 6.018, 'Nterm': 9.564, 'pKLys': 10.517, 'pKArg': 12.503}\n",
    "        pKCterm = IPC_score['Cterm']\n",
    "        pKAsp = IPC_score['pKAsp']\n",
    "        pKGlu = IPC_score['pKGlu']\n",
    "        pKCys = IPC_score['pKCys']\n",
    "        pKTyr = IPC_score['pKTyr']\n",
    "        pKHis = IPC_score['pk_his']\n",
    "        pKNterm = IPC_score['Nterm']\n",
    "        pKLys = IPC_score['pKLys'] \n",
    "        pKArg = IPC_score['pKArg']\n",
    "        pH = start_pH      \n",
    "        pHprev = 0.0         \n",
    "        pHnext = 14.0        \n",
    "        E = Epsilon  \n",
    "        temp = 0.01\n",
    "        nterm=peptide[0]\n",
    "        cterm=peptide[-1]\n",
    "    #will now cycle through all peptides until a pH within the epsilon is found       \n",
    "        while 1:             \n",
    "            QN1=-1.0/(1.0+pow(10,(pKCterm-pH)))                                        \n",
    "            QN2=-peptide.count('D')/(1.0+pow(10,(pKAsp-pH)))           \n",
    "            QN3=-peptide.count('E')/(1.0+pow(10,(pKGlu-pH)))           \n",
    "            QN4=-peptide.count('C')/(1.0+pow(10,(pKCys-pH)))           \n",
    "            QN5=-peptide.count('Y')/(1.0+pow(10,(pKTyr-pH)))        \n",
    "            QP1=peptide.count('H')/(1.0+pow(10,(pH-pKHis)))            \n",
    "            QP2=1.0/(1.0+pow(10,(pH-pKNterm)))                \n",
    "            QP3=peptide.count('K')/(1.0+pow(10,(pH-pKLys)))           \n",
    "            QP4=peptide.count('R')/(1.0+pow(10,(pH-pKArg)))            \n",
    "            NQ=QN1+QN2+QN3+QN4+QN5+QP1+QP2+QP3+QP4\n",
    "\n",
    "            if NQ<0.0:                                  \n",
    "                temp = pH\n",
    "                pH = pH-((pH-pHprev)/2.0)\n",
    "                pHnext = temp\n",
    "\n",
    "            else:\n",
    "                temp = pH\n",
    "                pH = pH + ((pHnext-pH)/2.0)\n",
    "                pHprev = temp\n",
    "            #terminal condition, finding pI with given precision defined by Epsilon\n",
    "            if (pH-pHprev<E) and (pHnext-pH<E): \n",
    "                return pH\n",
    "        \n",
    "    def Peptide_Neutral_pH(peptide):\n",
    "        z_dict = {'E': -1, 'D': -1, 'K': 1, 'R': 1} \n",
    "        charge = [z_dict.get(aa, 0.0) for aa in peptide]\n",
    "        spark=sum(charge)\n",
    "        return(spark)\n",
    "    \n",
    "    def Peptide_GRAVY(peptide):\n",
    "        hydro = {     \"A\": 1.800,\n",
    "        \"R\": -4.500,\n",
    "        \"N\": -3.500,\n",
    "        \"D\": -3.500,\n",
    "        \"C\": 2.500,\n",
    "        \"Q\": -3.500,\n",
    "        \"E\": -3.500,\n",
    "        \"G\": -0.400,\n",
    "        \"H\": -3.200,\n",
    "        \"I\": 4.500,\n",
    "        \"L\": 3.800,\n",
    "        \"K\": -3.900,\n",
    "        \"M\": 1.900,\n",
    "        \"F\": 2.800,\n",
    "        \"P\": -1.600,\n",
    "        \"S\": -0.800,\n",
    "        \"T\": -0.700,\n",
    "        \"W\": -0.900,\n",
    "        \"Y\": -1.300,\n",
    "        \"V\": 4.200,\n",
    "        }\n",
    "        hydro_list = [hydro.get(aa,0.0)for aa in peptide]\n",
    "        hydro_sum=sum(hydro_list)\n",
    "        gravy=hydro_sum/len(hydro_list)\n",
    "        return gravy \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Protein(object):\n",
    "    ''' Orginal Protein Information (i.e. Mass, Length, etc.). \n",
    "    Primary use is to compare to peptides following digestion.\n",
    "    =============================================================\n",
    "    Hyperparameters\n",
    "    ---------------\n",
    "    Peptide = A str(peptide sequence)\n",
    "    Length = Automatically caluclated based on length of Peptide.\n",
    "    Mass = Automatically caluclated based on mass of Peptide.\n",
    "    --------------------------------------------------------------\n",
    "    ''' \n",
    "    def __init__(self,df,target,identifier):\n",
    "        self.protein_peptide=df[[target, identifier]].set_index(identifier).to_dict()[target]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "Human=Protein(menu,\"Peptide\",\"Gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MLVRRGARAGPRMPRGWTALCLLSLLPSGFMSLDNNGTATPELPTQGTFSNVSTNVSYQETTTPSTLGSTSLHPVSQHGNEATTNITETTVKFTSTSVITSVYGNTNSSVQSQTSVISTVFTTPANVSTPETTLKPSLSPGNVSDLSTTSTSLATSPTKPYTSSSPILSDIKAEIKCSGIREVKLTQGICLEQNKTSSCAEFKKDRGEGLARVLCGEEQADADAGAQVCSLLLAQSEVRPQCLLLVLANRTEISSKLQLMKKHQSDLKKLGILDFTEQDVASHQSYSQKTLIALVTSGALLAVLGITGYFLMNRRSWSPTGERLGEDPYYTENGGGQGYSSGPGTSPEAQGKASVNRGAQENGTGQATSRNGHSARQHVVADTEL'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Human.protein_peptide[\"CD34\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MVSASGTSFFKGMLLGSISWVLITMFGQIHIRHRGQTQDHEHHHLRPPNRNDFLNTSKVILLELSKSIRVFCIIFGESEDESYWAVLKETWTKHCDKAELYDTKNDNLFNIESNDRWVQMRTAYKYVFEKYGDNYNWFFLALPTTFAVIENLKYLLFTRDASQPFYLGHTVIFGDLEYVTVEGGIVLSRELMKRLNRLLDNSETCADQSVIWKLSEDKQLAICLKYAGVHAENAEDYEGRDVFNTKPIAQLIEEALSNNPQQVVEGCCSDMAITFNGLTPQKMEVMMYGLYRLRAFGHYFNDTLVFLPPVGSEND'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence=menu.loc[7421,\"Peptide\"]\n",
    "sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    \n",
    "rules = {\n",
    "    'arg-c':         r'R',\n",
    "    'asp-n':         r'\\w(?=D)',\n",
    "    'bnps-skatole' : r'W',\n",
    "    'caspase 1':     r'(?<=[FWYL]\\w[HAT])D(?=[^PEDQKR])',\n",
    "    'caspase 2':     r'(?<=DVA)D(?=[^PEDQKR])',\n",
    "    'caspase 3':     r'(?<=DMQ)D(?=[^PEDQKR])',\n",
    "    'caspase 4':     r'(?<=LEV)D(?=[^PEDQKR])',\n",
    "    'caspase 5':     r'(?<=[LW]EH)D',\n",
    "    'caspase 6':     r'(?<=VE[HI])D(?=[^PEDQKR])',\n",
    "    'caspase 7':     r'(?<=DEV)D(?=[^PEDQKR])',\n",
    "    'caspase 8':     r'(?<=[IL]ET)D(?=[^PEDQKR])',\n",
    "    'caspase 9':     r'(?<=LEH)D',\n",
    "    'caspase 10':    r'(?<=IEA)D',\n",
    "    'chymotrypsin high specificity' : r'([FY](?=[^P]))|(W(?=[^MP]))',\n",
    "    'chymotrypsin low specificity':\n",
    "        r'([FLY](?=[^P]))|(W(?=[^MP]))|(M(?=[^PY]))|(H(?=[^DMPW]))',\n",
    "    'clostripain':   r'R',\n",
    "    'cnbr':          r'M',\n",
    "    'enterokinase':  r'(?<=[DE]{3})K',\n",
    "    'factor xa':     r'(?<=[AFGILTVM][DE]G)R',\n",
    "    'formic acid':   r'D',\n",
    "    'glutamyl endopeptidase': r'E',\n",
    "    'granzyme b':    r'(?<=IEP)D',\n",
    "    'hydroxylamine': r'N(?=G)',\n",
    "    'iodosobenzoic acid': r'W',\n",
    "    'lysc':          r'K',\n",
    "    'ntcb':          r'\\w(?=C)',\n",
    "    'pepsin ph1.3':  r'((?<=[^HKR][^P])[^R](?=[FL][^P]))|'\n",
    "                     r'((?<=[^HKR][^P])[FL](?=\\w[^P]))',\n",
    "    'pepsin ph2.0':  r'((?<=[^HKR][^P])[^R](?=[FLWY][^P]))|'\n",
    "                     r'((?<=[^HKR][^P])[FLWY](?=\\w[^P]))',\n",
    "    'proline endopeptidase': r'(?<=[HKR])P(?=[^P])',\n",
    "    'proteinase k':  r'[AEFILTVWY]',\n",
    "    'staphylococcal peptidase i': r'(?<=[^E])E',\n",
    "    'thermolysin':   r'[^DE](?=[AFILMV])',\n",
    "    'thrombin':      r'((?<=G)R(?=G))|'\n",
    "                     r'((?<=[AFGILTVM][AFGILTVWA]P)R(?=[^DE][^DE]))',\n",
    "    'trypsin':       r'([KR](?=[^P]))|((?<=W)K(?=P))|((?<=M)R(?=P))',\n",
    "    'trypsin_exception': r'((?<=[CD])K(?=D))|((?<=C)K(?=[HY]))|((?<=C)R(?=K))|((?<=R)R(?=[HR]))',\n",
    "}\n",
    "\n",
    "def update_rules(new_enzyme,new_rule,rules=rules):\n",
    "    rules.update(new_enzyme=new_rule)\n",
    "    return rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache(maxsize=None)\n",
    "def Cleaver(sequence, rule, missed_cleavages=0, min_length=None, max_length=None, exception=None):\n",
    "    peptides = []\n",
    "    if rule in rules:\n",
    "        rule = rules[rule]\n",
    "    exception = rules.get(exception, exception)\n",
    "    ml = missed_cleavages + 2\n",
    "    trange = range(ml) #returns range of 0 to ml-1\n",
    "    cleavage_sites = deque([0], maxlen=ml) # returns positions of cleavage sites between each other. \n",
    "    if min_length is None:\n",
    "        min_length = 1\n",
    "    cl = 1\n",
    "    if exception is not None: #locates postion of c-terminal by-product of cleavage. \n",
    "        exceptions = {x.end() for x in re.finditer(exception, sequence)}\n",
    "    for i in it.chain([x.end() for x in re.finditer(rule, sequence)],\n",
    "                   [None]):\n",
    "        if exception is not None and i in exceptions:\n",
    "            continue\n",
    "        cleavage_sites.append(i)\n",
    "#         print(cleavage_sites)\n",
    "        if cl < ml:\n",
    "            cl += 1\n",
    "        for j in trange[:cl - 1]:\n",
    "            seq = sequence[cleavage_sites[j]:cleavage_sites[-1]]\n",
    "            if (len(seq) >= min_length):\n",
    "                if (len(seq) <= max_length):\n",
    "                    peptides.append(seq)\n",
    "    return peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IRHRGQTQDHEHHH',\n",
       " 'LRPPNRNDF',\n",
       " 'FGESEDESYW',\n",
       " 'LKETWTKHCDK',\n",
       " 'AELYDTKNDN',\n",
       " 'IESNDRW',\n",
       " 'FEKYGDNYNW',\n",
       " 'FTRDASQP',\n",
       " 'LDNSETC',\n",
       " 'AEDYEGRDV',\n",
       " 'LSNNPQQ',\n",
       " 'VEGCCSDM']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missed_cleavages=0\n",
    "min_length=7\n",
    "max_length=100\n",
    "y=Cleaver(sequence,rule=\"thermolysin\",exception=None,missed_cleavages=0,min_length=7,max_length=100)\n",
    "y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defined Variables, Compiliers, Accessory Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Accessory Functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PEAKS_Importer(csv,drop_OG=True):\n",
    "    df=pd.read_csv(csv)\n",
    "    df[[\"Protein\",\"Y\"]]=df['Protein Accession'].str.split(\"|\",1,expand=True)\n",
    "    df[[\"Gene\",\"Species\"]]=df['Y'].str.split(\"_\",1,expand=True)\n",
    "    if drop_OG==True:\n",
    "        df.drop(columns=['Y', 'Protein Accession'], inplace=True)\n",
    "    else:\n",
    "        df.drop(columns=['Y'], inplace=True)\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Excel_Mapper(list_dfs, xls_path):\n",
    "    with ExcelWriter(xls_path) as writer:\n",
    "        for n, df in enumerate(list_dfs):\n",
    "            df.to_excel(writer,'sheet%s' % n)\n",
    "        writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def MeatWrapper(workbook,old,new,output):\n",
    "    meat = openpyxl.load_workbook(workbook)\n",
    "    wrap = meat[old]\n",
    "    wrap.title = new\n",
    "    meat.save(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Sampler(df,toothpicks,samples):\n",
    "    platter = {}\n",
    "    for toothpick, sample in zip(toothpicks, samples):\n",
    "        platter[sample] = df[toothpick]\n",
    "    return platter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Primary Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ButcherShop(df,target,identifier,rule, min_length=7,exception=None,max_length=100, pH=2.0, min_charge=2.0,missed=0):\n",
    "    pep_dict = {}\n",
    "    pep_dict_list = []\n",
    "    string_catcher=re.compile(r'^([A-Z]+)$')\n",
    "    print(f'You order is being processed and the butcher is preparing your {rule}-cut protein(s)!')\n",
    "    print(\"The butcher is working...\")\n",
    "    raw=df[[target, identifier]].set_index(identifier).to_dict()[target]\n",
    "    print(f\"Generating {rule}-cut peptides based on {missed}-missed cleavages. \")\n",
    "    for gene,peptide in raw.items():\n",
    "        pep_dict[gene] = Cleaver(peptide,rule=rule,min_length=min_length,exception=exception,missed_cleavages=missed, max_length=max_length)\n",
    "    for k, lst in pep_dict.items():    \n",
    "        cuts=len(lst)\n",
    "        for i in range(cuts):\n",
    "            d = {}\n",
    "            d.update({k: lst[i]})\n",
    "            d.update({'gene':k})\n",
    "            d.update({'aa_comp': dict(parser.amino_acid_composition(lst[i]))})\n",
    "            d.update({'peptide': lst[i]})\n",
    "            d.update({'Length': len(lst[i])})\n",
    "            d.update({'z': float(round(electrochem.charge(lst[i], pH=pH),3))})\n",
    "            d.update({'Mass': float(Scales.Mass(lst[i]))})\n",
    "            if d[\"z\"] > 0:\n",
    "                d.update({'m/z': d[\"Mass\"]/d['z']})\n",
    "            pep_dict_list.append(d)\n",
    "    print(\"Preparing your order...\")\n",
    "    pep_dict_list = [peptide for peptide in pep_dict_list if peptide['z'] >= float(min_charge)]\n",
    "    print(f'Order is up! You have acquired {len(pep_dict_list)} peptides that are between {min_length} and {max_length} amino acids!')\n",
    "    return pep_dict_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Deli(z,meat_package=False):\n",
    "    #z = list of dictionaries, keys must be equal thus will drop keys which are not cosistent between dictionaries\n",
    "    # use after ButcherShop\n",
    "    #returns dataframe\n",
    "    key_intersect = set(z[0].keys()).intersection(set(z[1].keys()))\n",
    "    zz = [{key:value for (key,value) in dicts.items() if key in key_intersect} for dicts in z]\n",
    "    ham = pd.DataFrame(zz)\n",
    "    if meat_package == True:\n",
    "        ham_counts=ham.groupby('gene').size().reset_index(name='counts')\n",
    "        ham=ham.merge(ham_counts,how='left', on=['gene'])\n",
    "#     ham.drop(ham.columns[0],axis=1,inplace=True)\n",
    "    columns=ham.columns.tolist()\n",
    "    r = re.compile(\"^[pP]\")\n",
    "    P = list(filter(r.match, columns)) \n",
    "    peptide =str(P[0])\n",
    "    ham[peptide]=ham[peptide].apply(\",\".join) # convert list within df to string\n",
    "    return ham"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You order is being processed and the butcher is preparing your trypsin-cut protein(s)!\n",
      "The butcher is working...\n",
      "Generating trypsin-cut peptides based on 0-missed cleavages. \n",
      "Preparing your order...\n",
      "Order is up! You have acquired 162838 peptides that are between 7 and 100 amino acids!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IGS23</th>\n",
       "      <th>gene</th>\n",
       "      <th>aa_comp</th>\n",
       "      <th>peptide</th>\n",
       "      <th>Length</th>\n",
       "      <th>z</th>\n",
       "      <th>Mass</th>\n",
       "      <th>m/z</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>IGS23</td>\n",
       "      <td>{'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...</td>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>64</td>\n",
       "      <td>2.642</td>\n",
       "      <td>6463.464677</td>\n",
       "      <td>2446.428720</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>IGS23</td>\n",
       "      <td>{'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...</td>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>64</td>\n",
       "      <td>2.642</td>\n",
       "      <td>6463.464677</td>\n",
       "      <td>2446.428720</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>IGS23</td>\n",
       "      <td>{'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...</td>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>64</td>\n",
       "      <td>2.642</td>\n",
       "      <td>6463.464677</td>\n",
       "      <td>2446.428720</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>IGS23</td>\n",
       "      <td>{'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...</td>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>64</td>\n",
       "      <td>2.642</td>\n",
       "      <td>6463.464677</td>\n",
       "      <td>2446.428720</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>IGS23</td>\n",
       "      <td>{'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...</td>\n",
       "      <td>QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...</td>\n",
       "      <td>64</td>\n",
       "      <td>2.642</td>\n",
       "      <td>6463.464677</td>\n",
       "      <td>2446.428720</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162833</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTC7A</td>\n",
       "      <td>{'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...</td>\n",
       "      <td>QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR</td>\n",
       "      <td>48</td>\n",
       "      <td>2.636</td>\n",
       "      <td>5134.516851</td>\n",
       "      <td>1947.844025</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162834</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTC7A</td>\n",
       "      <td>{'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...</td>\n",
       "      <td>QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR</td>\n",
       "      <td>48</td>\n",
       "      <td>2.636</td>\n",
       "      <td>5134.516851</td>\n",
       "      <td>1947.844025</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162835</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTC7A</td>\n",
       "      <td>{'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...</td>\n",
       "      <td>QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR</td>\n",
       "      <td>48</td>\n",
       "      <td>2.636</td>\n",
       "      <td>5134.516851</td>\n",
       "      <td>1947.844025</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162836</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTC7A</td>\n",
       "      <td>{'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...</td>\n",
       "      <td>QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR</td>\n",
       "      <td>48</td>\n",
       "      <td>2.636</td>\n",
       "      <td>5134.516851</td>\n",
       "      <td>1947.844025</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162837</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTC7A</td>\n",
       "      <td>{'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...</td>\n",
       "      <td>QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR</td>\n",
       "      <td>48</td>\n",
       "      <td>2.636</td>\n",
       "      <td>5134.516851</td>\n",
       "      <td>1947.844025</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>162838 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    IGS23   gene  \\\n",
       "0       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...  IGS23   \n",
       "1       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...  IGS23   \n",
       "2       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...  IGS23   \n",
       "3       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...  IGS23   \n",
       "4       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...  IGS23   \n",
       "...                                                   ...    ...   \n",
       "162833                                                NaN  TTC7A   \n",
       "162834                                                NaN  TTC7A   \n",
       "162835                                                NaN  TTC7A   \n",
       "162836                                                NaN  TTC7A   \n",
       "162837                                                NaN  TTC7A   \n",
       "\n",
       "                                                  aa_comp  \\\n",
       "0       {'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...   \n",
       "1       {'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...   \n",
       "2       {'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...   \n",
       "3       {'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...   \n",
       "4       {'Q': 3, 'L': 9, 'V': 2, 'S': 6, 'E': 4, 'P': ...   \n",
       "...                                                   ...   \n",
       "162833  {'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...   \n",
       "162834  {'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...   \n",
       "162835  {'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...   \n",
       "162836  {'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...   \n",
       "162837  {'Q': 5, 'S': 4, 'T': 2, 'C': 2, 'H': 1, 'E': ...   \n",
       "\n",
       "                                                  peptide  Length      z  \\\n",
       "0       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...      64  2.642   \n",
       "1       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...      64  2.642   \n",
       "2       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...      64  2.642   \n",
       "3       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...      64  2.642   \n",
       "4       QLVSEPVTISLPKPIMQPTEAEPMEPDPTLSLSGGSAIGLLAAGIL...      64  2.642   \n",
       "...                                                   ...     ...    ...   \n",
       "162833   QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR      48  2.636   \n",
       "162834   QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR      48  2.636   \n",
       "162835   QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR      48  2.636   \n",
       "162836   QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR      48  2.636   \n",
       "162837   QSTCHEAWQGLGEVLQAQGQNEAAVDCFLTALELEASSPVLPFSIIPR      48  2.636   \n",
       "\n",
       "               Mass          m/z  counts  \n",
       "0       6463.464677  2446.428720       7  \n",
       "1       6463.464677  2446.428720       7  \n",
       "2       6463.464677  2446.428720       7  \n",
       "3       6463.464677  2446.428720       7  \n",
       "4       6463.464677  2446.428720       7  \n",
       "...             ...          ...     ...  \n",
       "162833  5134.516851  1947.844025      48  \n",
       "162834  5134.516851  1947.844025      48  \n",
       "162835  5134.516851  1947.844025      48  \n",
       "162836  5134.516851  1947.844025      48  \n",
       "162837  5134.516851  1947.844025      48  \n",
       "\n",
       "[162838 rows x 9 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "z=ButcherShop(menu,target=\"Peptide\",identifier=\"Gene\", rule=\"trypsin\",missed=0,exception=None,min_length=7,max_length=100)\n",
    "zz=Deli(z,meat_package=True)\n",
    "zz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "meat=ButcherShop(menu,\"Peptide\",\"Gene\",\"thermolysin\")\n",
    "cuts=DeliShop(meat,meat_package=True)\n",
    "cuts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Origin(origin,target,Origin_Key=None,Target_Key=None,Origin_Label=None, Target_Label=None):\n",
    "    # primarily used to transfer whole protein measurments (i.e. aa_length) to data_frame with peptide products. Origin_Key is\n",
    "    # is used to find 'Merge' points (Target_Key) regardless of frequency (likely multiple peptides for one key/protein).Label is used\n",
    "    # capture information (peptide lenght, sequence, etc) from origin to target. New_Label is optional, default uses origin label to\n",
    "    # label newly created target column. \n",
    "    labeler=origin.set_index(Origin_Key).to_dict()[Origin_Label]\n",
    "    if Target_Label == None:\n",
    "        target[Origin_Label] = target[Target_Key].map(labeler)\n",
    "    else:\n",
    "        target[Target_Label] = target[Target_Key].map(labeler)\n",
    "    return target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>aa_comp</th>\n",
       "      <th>peptide</th>\n",
       "      <th>Length</th>\n",
       "      <th>z</th>\n",
       "      <th>Mass</th>\n",
       "      <th>m/z</th>\n",
       "      <th>counts</th>\n",
       "      <th>OG_LENGTH</th>\n",
       "      <th>% of Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>513795</th>\n",
       "      <td>KR193</td>\n",
       "      <td>{'M': 1, 'S': 8, 'Y': 19, 'G': 35, 'L': 3, 'C'...</td>\n",
       "      <td>MSYYGSYYGGLGYGCGGFGGLGYGYGCGCGSFRRLGSGCGYGGYGY...</td>\n",
       "      <td>81</td>\n",
       "      <td>4</td>\n",
       "      <td>8223</td>\n",
       "      <td>2055.750000</td>\n",
       "      <td>5</td>\n",
       "      <td>81</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30632</th>\n",
       "      <td>A0A0A0MT97</td>\n",
       "      <td>{'X': 1, 'T': 4, 'G': 2, 'W': 1, 'F': 2, 'K': ...</td>\n",
       "      <td>XTTGWFKIFAEGTKLIVTSP</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>2076</td>\n",
       "      <td>692.000000</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1599878</th>\n",
       "      <td>A0A075B6X0</td>\n",
       "      <td>{'X': 1, 'Y': 1, 'S': 1, 'G': 6, 'A': 1, 'D': ...</td>\n",
       "      <td>XYSGGGADGLTFGKGTHLIIQP</td>\n",
       "      <td>22</td>\n",
       "      <td>3</td>\n",
       "      <td>2071</td>\n",
       "      <td>690.333333</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129377</th>\n",
       "      <td>A0A075B709</td>\n",
       "      <td>{'X': 1, 'N': 1, 'S': 1, 'G': 4, 'Y': 1, 'Q': ...</td>\n",
       "      <td>XNSGGYQKVTFGTGTKLQVIP</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>2078</td>\n",
       "      <td>692.666667</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1599881</th>\n",
       "      <td>A0A096LPG2</td>\n",
       "      <td>{'P': 3, 'T': 1, 'L': 4, 'C': 3, 'S': 3, 'F': ...</td>\n",
       "      <td>PTLCCLSCFSSLWPRWEDHLRPG</td>\n",
       "      <td>23</td>\n",
       "      <td>4</td>\n",
       "      <td>2683</td>\n",
       "      <td>670.750000</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267664</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'S': 1, 'A': 1, 'T': 1, 'V': 2, 'L': 1, 'K': 1}</td>\n",
       "      <td>SATVLVK</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>698</td>\n",
       "      <td>349.000000</td>\n",
       "      <td>9509</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266252</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'V': 1, 'K': 2, 'A': 1, 'Y': 1, 'N': 1, 'E': 1}</td>\n",
       "      <td>VKAYNEK</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>831</td>\n",
       "      <td>277.000000</td>\n",
       "      <td>9509</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262205</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'E': 1, 'K': 2, 'P': 3, 'A': 1}</td>\n",
       "      <td>EKPPPAK</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>746</td>\n",
       "      <td>248.666667</td>\n",
       "      <td>9509</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266254</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'A': 1, 'Y': 1, 'N': 1, 'E': 1, 'K': 2, 'G': 1}</td>\n",
       "      <td>AYNEKGK</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>789</td>\n",
       "      <td>263.000000</td>\n",
       "      <td>9509</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259322</th>\n",
       "      <td>TITIN</td>\n",
       "      <td>{'H': 1, 'S': 2, 'P': 2, 'V': 1, 'R': 1}</td>\n",
       "      <td>HSPSPVR</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>760</td>\n",
       "      <td>253.333333</td>\n",
       "      <td>9509</td>\n",
       "      <td>34350</td>\n",
       "      <td>0.020378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2683232 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene                                            aa_comp  \\\n",
       "513795        KR193  {'M': 1, 'S': 8, 'Y': 19, 'G': 35, 'L': 3, 'C'...   \n",
       "30632    A0A0A0MT97  {'X': 1, 'T': 4, 'G': 2, 'W': 1, 'F': 2, 'K': ...   \n",
       "1599878  A0A075B6X0  {'X': 1, 'Y': 1, 'S': 1, 'G': 6, 'A': 1, 'D': ...   \n",
       "129377   A0A075B709  {'X': 1, 'N': 1, 'S': 1, 'G': 4, 'Y': 1, 'Q': ...   \n",
       "1599881  A0A096LPG2  {'P': 3, 'T': 1, 'L': 4, 'C': 3, 'S': 3, 'F': ...   \n",
       "...             ...                                                ...   \n",
       "267664        TITIN   {'S': 1, 'A': 1, 'T': 1, 'V': 2, 'L': 1, 'K': 1}   \n",
       "266252        TITIN   {'V': 1, 'K': 2, 'A': 1, 'Y': 1, 'N': 1, 'E': 1}   \n",
       "262205        TITIN                   {'E': 1, 'K': 2, 'P': 3, 'A': 1}   \n",
       "266254        TITIN   {'A': 1, 'Y': 1, 'N': 1, 'E': 1, 'K': 2, 'G': 1}   \n",
       "259322        TITIN           {'H': 1, 'S': 2, 'P': 2, 'V': 1, 'R': 1}   \n",
       "\n",
       "                                                   peptide  Length  z  Mass  \\\n",
       "513795   MSYYGSYYGGLGYGCGGFGGLGYGYGCGCGSFRRLGSGCGYGGYGY...      81  4  8223   \n",
       "30632                                 XTTGWFKIFAEGTKLIVTSP      20  3  2076   \n",
       "1599878                             XYSGGGADGLTFGKGTHLIIQP      22  3  2071   \n",
       "129377                               XNSGGYQKVTFGTGTKLQVIP      21  3  2078   \n",
       "1599881                            PTLCCLSCFSSLWPRWEDHLRPG      23  4  2683   \n",
       "...                                                    ...     ... ..   ...   \n",
       "267664                                             SATVLVK       7  2   698   \n",
       "266252                                             VKAYNEK       7  3   831   \n",
       "262205                                             EKPPPAK       7  3   746   \n",
       "266254                                             AYNEKGK       7  3   789   \n",
       "259322                                             HSPSPVR       7  3   760   \n",
       "\n",
       "                 m/z  counts  OG_LENGTH  % of Protein  \n",
       "513795   2055.750000       5         81    100.000000  \n",
       "30632     692.000000       5         20    100.000000  \n",
       "1599878   690.333333       3         22    100.000000  \n",
       "129377    692.666667       5         21    100.000000  \n",
       "1599881   670.750000       3         23    100.000000  \n",
       "...              ...     ...        ...           ...  \n",
       "267664    349.000000    9509      34350      0.020378  \n",
       "266252    277.000000    9509      34350      0.020378  \n",
       "262205    248.666667    9509      34350      0.020378  \n",
       "266254    263.000000    9509      34350      0.020378  \n",
       "259322    253.333333    9509      34350      0.020378  \n",
       "\n",
       "[2683232 rows x 10 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fresh=Peptide_Origin(menu,z,Origin_Key=\"Gene\",Target_Key=\"gene\",Origin_Label=\"Protein_Length\",Target_Label=\"OG_LENGTH\")\n",
    "fresh[\"% of Protein\"]= (fresh[\"Length\"]/fresh[\"OG_LENGTH\"])*100\n",
    "fresh.sort_values(by=[\"% of Protein\"],ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICE_cloud(df,x=0,target=\"aa_comp\"):\n",
    "    word_could_dict = df[target][x]\n",
    "    wordcloud = WordCloud(width = 500, height = 500,scale=10,prefer_horizontal=1,relative_scaling=1,min_font_size=18,max_font_size=48,font_step=8,\n",
    "                      background_color='white',contour_width=1,contour_color=\"black\" ).generate_from_frequencies(word_could_dict)\n",
    "    plt.figure(figsize=(15,8))\n",
    "    plt.imshow(wordcloud)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Handles Up to 3 Replicates (t_id1-3) per Selection, used for removing amino acids from N-terminal during PEAKS exports with enzyme that cleave at C-terminal. \n",
    "#Can be used to remove M from N-terminal of peptides produced by enzymes which cleave at N-terminal of target aa. \n",
    "def Butcher(df,ident1=None,ident2=None,ident3=None,t_id1=None,t_id2=None,t_id3=None,t_value=0,acid=[\"J\",\"Z\"],labels=list(),excel_mapper=True,excel_name=None):\n",
    "    raw=df.loc[:,df.columns.str.contains(ident1)]\n",
    "    tag=df[labels]\n",
    "    raw = pd.concat([raw, tag], axis=1)\n",
    "    raw[\"Peptide\"]= raw[\"Peptide\"].str.replace('\\W+',\"\")\n",
    "    raw[\"Peptide\"]= raw[\"Peptide\"].str.replace('\\d+',\"\")\n",
    "    raw[\"Peptide\"]= raw[\"Peptide\"].apply(lambda x : x[1:] if x.startswith(tuple(acid)) else x)\n",
    "    cuts=raw.loc[:,raw.columns.str.contains(ident2)]\n",
    "    if ident3 != None:\n",
    "        cuts=cuts.loc[:,cuts.columns.str.contains(ident3)]\n",
    "    cuts = pd.concat([cut, tag], axis=1)\n",
    "    blade = cuts.filter(regex=r'^AREA').isin(['0']).all(axis=1)\n",
    "    cuts=cuts.loc[~blade]\n",
    "    cuts.reset_index(inplace=True)\n",
    "    excels=[raw,cuts]\n",
    "    Excel_Mapper(excels,excel_name +\".xlsx\")\n",
    "    return raw, cuts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICE_bar(df,x=0,target=None,lock=None,key=None):\n",
    "    test_dict=df.loc[df[lock] == key][target].iloc[x]\n",
    "    x=list(test_dict.keys())\n",
    "    y=list(test_dict.values())\n",
    "    plt.bar(x,y, color='g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICE_plot(df,x=0,target=None,lock=None,key=None):\n",
    "    test_dict=df.loc[df[lock] == key][target].iloc[x]\n",
    "    x=list(test_dict.keys())\n",
    "    y=list(test_dict.values())\n",
    "    plt.plot(x,y, color='g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "single positional indexer is out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-61-4ed3b24bd64f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mPeptide_ICE_plot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfresh\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtarget\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"aa_comp\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlock\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"gene\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"LHDA\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-60-8374713b45df>\u001b[0m in \u001b[0;36mPeptide_ICE_plot\u001b[1;34m(df, x, target, lock, key)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mPeptide_ICE_plot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtarget\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mlock\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m     \u001b[0mtest_dict\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mlock\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mtarget\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m     \u001b[0mx\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtest_dict\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m     \u001b[0my\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtest_dict\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mplt\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcolor\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'g'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\SciFly\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   1766\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1767\u001b[0m             \u001b[0mmaybe_callable\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1768\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmaybe_callable\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1769\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1770\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\SciFly\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   2136\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2137\u001b[0m             \u001b[1;31m# validate the location\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2138\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_validate_integer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2139\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2140\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\SciFly\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_validate_integer\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   2061\u001b[0m         \u001b[0mlen_axis\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2062\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mkey\u001b[0m \u001b[1;33m>=\u001b[0m \u001b[0mlen_axis\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mkey\u001b[0m \u001b[1;33m<\u001b[0m \u001b[1;33m-\u001b[0m\u001b[0mlen_axis\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2063\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"single positional indexer is out-of-bounds\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2064\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2065\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_getitem_tuple\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: single positional indexer is out-of-bounds"
     ]
    }
   ],
   "source": [
    "Peptide_ICE_plot(fresh,x=0,target=\"aa_comp\", lock=\"gene\",key=\"LHDA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Wash(peptide,ccut=None, PEAKS=False):\n",
    "    if ccut is None:\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "    elif ccut == \"Trypsin\":\n",
    "        x=[ 'R', 'K']\n",
    "        if PEAKS:\n",
    "            peptide= peptide.apply(lambda x : x[1:] if x.startswith(tuple(x)) else x)\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "        return(peptide)\n",
    "    elif ccut == \"AspN\":\n",
    "        x=['A', 'R', 'N','D', 'C', 'Q','E', 'G', 'H',\n",
    "             'I', 'L', 'K', 'F', 'P',\n",
    "             'S', 'T', 'W','Y', 'V']\n",
    "        if PEAKS:\n",
    "            peptide= peptide.apply(lambda x : x[1:] if x.startswith(tuple(x)) else x)\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "        return(peptide)\n",
    "    elif ccut == \"GluC\":\n",
    "        x=['E']\n",
    "        if PEAKS:\n",
    "            peptide= peptide.apply(lambda x : x[1:] if x.startswith(tuple(x)) else x)\n",
    "        peptide = peptide.str.replace('\\W+','')\n",
    "        peptide = peptide.str.replace('\\d+','')\n",
    "        return(peptide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_ICEmaker(peptide,excel_name=\"Test\",freeze_it=0,cut_it=6):\n",
    "    ice=peptide.str.slice(freeze_it,cut_it)\n",
    "    ice_ice = ice.str.slice(freeze_it,cut_it)\n",
    "    ice_ice.to_excel('ICE '+excel_name+\".xlsx\")\n",
    "    return ice_ice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache(maxsize=None)\n",
    "def Pep2Pro(protein,peptides):\n",
    "    protein = re.sub(r'[^A-Z]', '', protein)\n",
    "    mask = np.zeros(len(protein), dtype=np.int8)\n",
    "    for peptide in peptides:\n",
    "        indices = [m.start() for m in re.finditer(\n",
    "            '(?={})'.format(re.sub(r'[^A-Z]', '', peptide)), protein)]\n",
    "        for i in indices:\n",
    "            mask[i:i + len(peptide)] = 1\n",
    "    return mask.sum(dtype=float) / mask.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Peptide_Inspector(df,target,new_column=None):\n",
    "    all_data = []\n",
    "    for peptide in df[target]:\n",
    "        temp=dict(parser.amino_acid_composition(peptide))\n",
    "        all_data.append(temp)\n",
    "    df[new_column]=all_data\n",
    "    print(f\"Peptide inspection is completed! Amino acid dictionaries are stored under {new_column} in dataframe!\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Marinate (df,target,length,IPC=False,Hydro=False,GRAVY=False,NeutralZ=False,Peptide_Inspector=False):\n",
    "    print(\"Marinating peptides...\")\n",
    "    if IPC is True:\n",
    "        df[\"IPC\"]=df[target].apply(Scales.Peptide_IPC)\n",
    "        print(\"IPC calculated!\")\n",
    "    if GRAVY is True:\n",
    "        df[\"Hydro_Sum\"]=df[target].apply(Scales.Peptide_GRAVY)\n",
    "        print(\"Its all GRAVY Baby!\")\n",
    "    if NeutralZ is True:\n",
    "        df[\"Neutral_Z\"]=df[target].apply(Scales.Peptide_Neutral_pH)\n",
    "        print(\"Charge at Neutral pH added!\")\n",
    "    print(\"Peptides have been marinated!\")\n",
    "    return df\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Wishbone(df,bone,split):\n",
    "    x=df[bone].value_counts()\n",
    "    x2=pd.DataFrame(x,columns=[bone])\n",
    "    x3= x2.loc[(x2[bone])>int(split)]\n",
    "    x3z=len(x3)\n",
    "    x4=x3z/len(x2)*100\n",
    "    print(f\"The number of proteins with > {split} peptides : %.1f\" % x3z)\n",
    "    print(f\"Ratio of Proteins with > {split} Peptides Identified: %.3f\" % x4)\n",
    "    return x3,x3z,x4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Sweet_N_Sour(set1,set2):\n",
    "    common_IDs=list(set(set1)&set(set2))\n",
    "    common=len(common_IDs)\n",
    "    unique=len(list(set(set1)^set(set2)))\n",
    "    unique_set1=list(set(set1)-set(set2))\n",
    "    unique_set2=list(set(set2)-set(set1))\n",
    "    set1_count=len(unique_set1)\n",
    "    set2_count=len(unique_set2)\n",
    "    Ratio=(unique/(common+unique))*100\n",
    "    print(f\"The number of common peptides is: %.3f\" % common)\n",
    "    print(f\"The number of unique peptides in set1 is: %.3f\" % set1_count)\n",
    "    print(f\"The number of unique peptides in set1 is: %.3f\" % set2_count)\n",
    "    print(f\"The Ratio of Unique to Common proteins is: %.3f\" % Ratio)\n",
    "    return common_IDs,unique_set1, unique_set2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CorrPie(df,apples,figsize=(16,8),cmap='Blues',title=\"Title\",fontsize=18,pad=16,save_name=\"Heat_Test\",dpi=600,fmt='eps'):\n",
    "    og_corr=df[apples]\n",
    "    plt.figure(figsize=figsize)\n",
    "    mask=np.triu(np.ones_like(og_corr.corr(), dtype=np.bool))\n",
    "    heatmap=heatmap = sns.heatmap(og_corr.corr(), mask=mask, vmin=-1, vmax=1, annot=True, cmap=cmap)\n",
    "    heatmap.set_title(title, fontdict={'fontsize':fontsize}, pad=pad)\n",
    "    plt.savefig(save_name,format=fmt,dpi=dpi,bbox_inches=\"tight\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def WeddingCake(df,x,y,z,dpi=300,s=12,alpha=0.3,edgecolor='k',color=\"blue\",my_viewx=20,my_viewy=50,\n",
    "               xlabel=\"Parameter xlabel\",ylabel=\"Parameter ylabel\",zlabel=\"Parameter zlabel\",fmt='png',\n",
    "               figx=10,figy=10,xmin=0,xmax=1000,ymin=0,ymax=1000,zmin=0,zmax=1000):\n",
    "    fig = plt.figure(figsize=(figx,figy),dpi=dpi)\n",
    "    ax = fig.gca(projection='3d')\n",
    "    X = df[x]\n",
    "    Y = df[y]\n",
    "    Z = df[z]\n",
    "    ax.scatter(X,Y,Z,color=color,s=s,alpha=alpha,edgecolor=edgecolor)\n",
    "    ax.view_init(my_viewx,my_viewy)\n",
    "    ax.set_xlabel(xlabel)\n",
    "    ax.set_ylabel(ylabel)\n",
    "    ax.set_zlabel(zlabel)\n",
    "    ax.xaxis._axinfo[\"grid\"].update({\"linewidth\":1, \"color\" : \"grey\"})\n",
    "    ax.yaxis._axinfo[\"grid\"].update({\"linewidth\":1, \"color\" : \"grey\"})\n",
    "    ax.zaxis._axinfo[\"grid\"]['color'] = \"k\"\n",
    "    ax.zaxis._axinfo[\"grid\"]['linestyle'] = \"--\"\n",
    "    ax.set_ylim3d(ymin,ymax)\n",
    "    ax.set_xlim3d(xmin,xmax)\n",
    "    ax.set_zlim3d(zmin,zmax)\n",
    "    FigTitle=input('Figure Title:')\n",
    "    plt.savefig(FigTitle,format=fmt,dpi=dpi,bbox_inches=\"tight\")\n",
    "    plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Cake(df,x,y,dpi=600,s=25,alpha=0.25,edgecolor='k',color=\"blue\",fmt='eps',labsize=18,\n",
    "               figx=10,figy=10,xmin=0,xmax=1000,ymin=0,ymax=1000,xlabel=\"xlabel parameter\",ylabel=\"ylabel parameter\",\n",
    "        loc=2,pad=1,borderpad=1,frameon=True, show=False):\n",
    "    fig, ax = plt.subplots(1,1,figsize=(figx,figy),dpi=dpi)\n",
    "    X = df[x]\n",
    "    Y = df[y]\n",
    "    plt.scatter(X,Y,color=color,s=s, alpha=alpha,edgecolor=edgecolor)\n",
    "    plt.xlabel(xlabel, fontsize=labsize)\n",
    "    plt.ylabel(ylabel, fontsize=labsize)\n",
    "    corr, _ = pearsonr(X, Y)\n",
    "    # loc works the same as it does with figures (though best doesn't work)\n",
    "    # pad=5 will increase the size of padding between the border and text\n",
    "    # borderpad=5 will increase the distance between the border and the axes\n",
    "    # frameon=False will remove the box around the text\n",
    "    anchored_text = AnchoredText('Pearsons correlation: r = %.3f' % corr, loc=loc,pad=pad,borderpad=borderpad,frameon=frameon)\n",
    "    ax.add_artist(anchored_text)\n",
    "    FigTitle=input('Figure Title:')\n",
    "    plt.savefig(FigTitle,format=fmt,dpi=dpi,bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CakePop(df,x,xi,y,yi, figx=10,figy=10,dpi=600,font_scale=1.5,figstyle=\"white\",xlabel=\"xlabel\",ylabel=\"ylabel\",labsize=18,\n",
    "           cmapx=-0.3,cmapy=0.0,sizex=10,sizey=200,loc='upper left',pad=1,borderpad=1,frameon=True):\n",
    "    sns.set(font_scale=font_scale)\n",
    "    cake=sns.set_style(figstyle)\n",
    "    cmap = sns.cubehelix_palette(start=cmapx, rot=cmapy, as_cmap=True)\n",
    "    X = df[x]\n",
    "    Y = df[y]\n",
    "    cake=sns.relplot(data=df,\n",
    "    x=X, y=Y,\n",
    "    hue=xi, size=yi,\n",
    "    palette=cmap, sizes=(sizex, sizey))\n",
    "    ax = cake.axes[0,0]\n",
    "    ####\n",
    "    corr, _ = pearsonr(X, Y)\n",
    "    anchored_text = AnchoredText('Pearsons correlation: r = %.3f' % corr, loc=loc, prop=dict(size=labsize*0.5),pad=pad,borderpad=borderpad,frameon=frameon)\n",
    "    ax.add_artist(anchored_text)\n",
    "    ####\n",
    "    plt.xlabel(xlabel, fontsize=labsize)\n",
    "    plt.ylabel(ylabel, fontsize=labsize)\n",
    "    FigTitle=input('Figure Title:')\n",
    "    plt.savefig(FigTitle,format=fmt,dpi=dpi,bbox_inches=\"tight\")\n",
    "    plt.show(cake)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
